
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://eko-refugium.github.io/DNAbyte/reference/">
      
      
        <link rel="prev" href="../concepts/">
      
      
      
        
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.0">
    
    
      
        <title>Reference - DNAbyte Documentation</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.618322db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#api-reference" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="DNAbyte Documentation" class="md-header__button md-logo" aria-label="DNAbyte Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            DNAbyte Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Reference
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/Eko-Refugium/DNAbyte" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    Eko-Refugium/DNAbyte
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="DNAbyte Documentation" class="md-nav__button md-logo" aria-label="DNAbyte Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    DNAbyte Documentation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/Eko-Refugium/DNAbyte" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    Eko-Refugium/DNAbyte
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../usage/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Usage
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../simulation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Simulation
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../concepts/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Concepts
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    
  
    Reference
  

    
  </span>
  
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    
  
    Reference
  

    
  </span>
  
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#data-classes" class="md-nav__link">
    <span class="md-ellipsis">
      
        Data Classes
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Data Classes">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#data-base-class" class="md-nav__link">
    <span class="md-ellipsis">
      
        Data (Base Class)
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.base.Data" class="md-nav__link">
    <span class="md-ellipsis">
      
        Data
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.base.Data.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __init__
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.base.Data.__str__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __str__
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.base.Data.calculate_total_bytes" class="md-nav__link">
    <span class="md-ellipsis">
      
        calculate_total_bytes
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.base.Data.from_folder" class="md-nav__link">
    <span class="md-ellipsis">
      
        from_folder
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="from_folder">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#binarycode" class="md-nav__link">
    <span class="md-ellipsis">
      
        BinaryCode
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.binarycode.BinaryCode" class="md-nav__link">
    <span class="md-ellipsis">
      
        BinaryCode
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.binarycode.BinaryCode.__getitem__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __getitem__
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.binarycode.BinaryCode.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __init__
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.binarycode.BinaryCode.__iter__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __iter__
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.binarycode.BinaryCode.__len__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __len__
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.binarycode.BinaryCode.__repr__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __repr__
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.binarycode.BinaryCode.compare" class="md-nav__link">
    <span class="md-ellipsis">
      
        compare
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.binarycode.BinaryCode.compare_binary_strings" class="md-nav__link">
    <span class="md-ellipsis">
      
        compare_binary_strings
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.binarycode.BinaryCode.random" class="md-nav__link">
    <span class="md-ellipsis">
      
        random
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="random">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#nucleobasecode" class="md-nav__link">
    <span class="md-ellipsis">
      
        NucleobaseCode
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.nucleobasecode.NucleobaseCode" class="md-nav__link">
    <span class="md-ellipsis">
      
        NucleobaseCode
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.nucleobasecode.NucleobaseCode.__getitem__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __getitem__
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.nucleobasecode.NucleobaseCode.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __init__
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.nucleobasecode.NucleobaseCode.__len__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __len__
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.nucleobasecode.NucleobaseCode.get_codeword" class="md-nav__link">
    <span class="md-ellipsis">
      
        get_codeword
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.nucleobasecode.NucleobaseCode.random" class="md-nav__link">
    <span class="md-ellipsis">
      
        random
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.nucleobasecode.NucleobaseCode.validate" class="md-nav__link">
    <span class="md-ellipsis">
      
        validate
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="validate">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#insilicodna" class="md-nav__link">
    <span class="md-ellipsis">
      
        InSilicoDNA
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.insilicodna.InSilicoDNA" class="md-nav__link">
    <span class="md-ellipsis">
      
        InSilicoDNA
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.__getitem__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __getitem__
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __init__
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.__iter__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __iter__
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.__len__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __len__
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.add_sequence" class="md-nav__link">
    <span class="md-ellipsis">
      
        add_sequence
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.get_nucleotide_counts" class="md-nav__link">
    <span class="md-ellipsis">
      
        get_nucleotide_counts
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.get_sequence" class="md-nav__link">
    <span class="md-ellipsis">
      
        get_sequence
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.get_sequence_lengths" class="md-nav__link">
    <span class="md-ellipsis">
      
        get_sequence_lengths
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.random" class="md-nav__link">
    <span class="md-ellipsis">
      
        random
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="random">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.random--parameters" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters:
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.random--returns" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns:
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.remove_sequence" class="md-nav__link">
    <span class="md-ellipsis">
      
        remove_sequence
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.validate" class="md-nav__link">
    <span class="md-ellipsis">
      
        validate
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#core-modules" class="md-nav__link">
    <span class="md-ellipsis">
      
        Core Modules
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Core Modules">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#params" class="md-nav__link">
    <span class="md-ellipsis">
      
        Params
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.params.Params" class="md-nav__link">
    <span class="md-ellipsis">
      
        Params
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.params.Params.params_range" class="md-nav__link">
    <span class="md-ellipsis">
      
        params_range
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="params_range">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#library" class="md-nav__link">
    <span class="md-ellipsis">
      
        Library
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.library.Library" class="md-nav__link">
    <span class="md-ellipsis">
      
        Library
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.library.Library.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __init__
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.library.Library.motive_pairs" class="md-nav__link">
    <span class="md-ellipsis">
      
        motive_pairs
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.library.Library.motive_pairs_poly" class="md-nav__link">
    <span class="md-ellipsis">
      
        motive_pairs_poly
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.library.Library.read_library" class="md-nav__link">
    <span class="md-ellipsis">
      
        read_library
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="read_library">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#oligo" class="md-nav__link">
    <span class="md-ellipsis">
      
        Oligo
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.oligo.Oligo" class="md-nav__link">
    <span class="md-ellipsis">
      
        Oligo
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.oligo.Oligo.mirror" class="md-nav__link">
    <span class="md-ellipsis">
      
        mirror
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="mirror">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#oligopool" class="md-nav__link">
    <span class="md-ellipsis">
      
        OligoPool
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.oligopool.OligoPool" class="md-nav__link">
    <span class="md-ellipsis">
      
        OligoPool
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#data-classes" class="md-nav__link">
    <span class="md-ellipsis">
      
        Data Classes
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Data Classes">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#data-base-class" class="md-nav__link">
    <span class="md-ellipsis">
      
        Data (Base Class)
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.base.Data" class="md-nav__link">
    <span class="md-ellipsis">
      
        Data
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.base.Data.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __init__
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.base.Data.__str__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __str__
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.base.Data.calculate_total_bytes" class="md-nav__link">
    <span class="md-ellipsis">
      
        calculate_total_bytes
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.base.Data.from_folder" class="md-nav__link">
    <span class="md-ellipsis">
      
        from_folder
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="from_folder">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#binarycode" class="md-nav__link">
    <span class="md-ellipsis">
      
        BinaryCode
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.binarycode.BinaryCode" class="md-nav__link">
    <span class="md-ellipsis">
      
        BinaryCode
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.binarycode.BinaryCode.__getitem__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __getitem__
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.binarycode.BinaryCode.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __init__
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.binarycode.BinaryCode.__iter__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __iter__
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.binarycode.BinaryCode.__len__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __len__
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.binarycode.BinaryCode.__repr__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __repr__
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.binarycode.BinaryCode.compare" class="md-nav__link">
    <span class="md-ellipsis">
      
        compare
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.binarycode.BinaryCode.compare_binary_strings" class="md-nav__link">
    <span class="md-ellipsis">
      
        compare_binary_strings
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.binarycode.BinaryCode.random" class="md-nav__link">
    <span class="md-ellipsis">
      
        random
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="random">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#nucleobasecode" class="md-nav__link">
    <span class="md-ellipsis">
      
        NucleobaseCode
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.nucleobasecode.NucleobaseCode" class="md-nav__link">
    <span class="md-ellipsis">
      
        NucleobaseCode
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.nucleobasecode.NucleobaseCode.__getitem__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __getitem__
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.nucleobasecode.NucleobaseCode.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __init__
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.nucleobasecode.NucleobaseCode.__len__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __len__
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.nucleobasecode.NucleobaseCode.get_codeword" class="md-nav__link">
    <span class="md-ellipsis">
      
        get_codeword
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.nucleobasecode.NucleobaseCode.random" class="md-nav__link">
    <span class="md-ellipsis">
      
        random
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.nucleobasecode.NucleobaseCode.validate" class="md-nav__link">
    <span class="md-ellipsis">
      
        validate
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="validate">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#insilicodna" class="md-nav__link">
    <span class="md-ellipsis">
      
        InSilicoDNA
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.insilicodna.InSilicoDNA" class="md-nav__link">
    <span class="md-ellipsis">
      
        InSilicoDNA
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.__getitem__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __getitem__
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __init__
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.__iter__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __iter__
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.__len__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __len__
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.add_sequence" class="md-nav__link">
    <span class="md-ellipsis">
      
        add_sequence
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.get_nucleotide_counts" class="md-nav__link">
    <span class="md-ellipsis">
      
        get_nucleotide_counts
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.get_sequence" class="md-nav__link">
    <span class="md-ellipsis">
      
        get_sequence
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.get_sequence_lengths" class="md-nav__link">
    <span class="md-ellipsis">
      
        get_sequence_lengths
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.random" class="md-nav__link">
    <span class="md-ellipsis">
      
        random
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="random">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.random--parameters" class="md-nav__link">
    <span class="md-ellipsis">
      
        Parameters:
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.random--returns" class="md-nav__link">
    <span class="md-ellipsis">
      
        Returns:
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.remove_sequence" class="md-nav__link">
    <span class="md-ellipsis">
      
        remove_sequence
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.validate" class="md-nav__link">
    <span class="md-ellipsis">
      
        validate
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#core-modules" class="md-nav__link">
    <span class="md-ellipsis">
      
        Core Modules
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Core Modules">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#params" class="md-nav__link">
    <span class="md-ellipsis">
      
        Params
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.params.Params" class="md-nav__link">
    <span class="md-ellipsis">
      
        Params
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.params.Params.params_range" class="md-nav__link">
    <span class="md-ellipsis">
      
        params_range
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="params_range">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#library" class="md-nav__link">
    <span class="md-ellipsis">
      
        Library
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.library.Library" class="md-nav__link">
    <span class="md-ellipsis">
      
        Library
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.library.Library.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      
        __init__
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.library.Library.motive_pairs" class="md-nav__link">
    <span class="md-ellipsis">
      
        motive_pairs
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.library.Library.motive_pairs_poly" class="md-nav__link">
    <span class="md-ellipsis">
      
        motive_pairs_poly
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.library.Library.read_library" class="md-nav__link">
    <span class="md-ellipsis">
      
        read_library
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="read_library">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#oligo" class="md-nav__link">
    <span class="md-ellipsis">
      
        Oligo
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.oligo.Oligo" class="md-nav__link">
    <span class="md-ellipsis">
      
        Oligo
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.oligo.Oligo.mirror" class="md-nav__link">
    <span class="md-ellipsis">
      
        mirror
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="mirror">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#oligopool" class="md-nav__link">
    <span class="md-ellipsis">
      
        OligoPool
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dnabyte.oligopool.OligoPool" class="md-nav__link">
    <span class="md-ellipsis">
      
        OligoPool
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="api-reference">API Reference<a class="headerlink" href="#api-reference" title="Permanent link">&para;</a></h1>
<h2 id="data-classes">Data Classes<a class="headerlink" href="#data-classes" title="Permanent link">&para;</a></h2>
<p>DNAbyte uses different data structures to represent information at various stages of the DNA storage pipeline.</p>
<h3 id="data-base-class">Data (Base Class)<a class="headerlink" href="#data-base-class" title="Permanent link">&para;</a></h3>


<div class="doc doc-object doc-class">



<a id="dnabyte.data_classes.base.Data"></a>
    <div class="doc doc-contents first">



        <p>Data is the base class for all data types in the DNAbyte pipeline.</p>
<p>It can be instantiated 
 - with a list of absolute paths to files or, alternatively
 - with a path to a folder using the alternative constructor from_folder.</p>
<p>:param file_paths: A list of absolute paths to files.</p>








              <details class="mkdocstrings-source">
                <summary>Source code in <code>dnabyte/data_classes/base.py</code></summary>
                <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span>
<span class="normal">95</span>
<span class="normal">96</span>
<span class="normal">97</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">Data</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Data is the base class for all data types in the DNAbyte pipeline.</span>

<span class="sd">    It can be instantiated </span>
<span class="sd">     - with a list of absolute paths to files or, alternatively</span>
<span class="sd">     - with a path to a folder using the alternative constructor from_folder.</span>

<span class="sd">    :param file_paths: A list of absolute paths to files.    </span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Class constant for maximum file size (1MB in bytes)</span>
    <span class="n">MAX_FILE_SIZE</span> <span class="o">=</span> <span class="mi">1024</span> <span class="o">*</span> <span class="mi">1024</span>  <span class="c1"># 1MB</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_paths</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialize Data object with file paths.</span>

<span class="sd">        :param file_paths: List of absolute paths to files</span>
<span class="sd">        :raises TypeError: If file_paths is not a list</span>
<span class="sd">        :raises ValueError: If file_paths is empty, contains invalid paths, or files exceed size limit</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Validate input type</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">file_paths</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;file_paths must be a list, got </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">file_paths</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Check if list is empty</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">file_paths</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;file_paths cannot be empty&quot;</span><span class="p">)</span>

        <span class="c1"># Validate each file path</span>
        <span class="k">for</span> <span class="n">file_path</span> <span class="ow">in</span> <span class="n">file_paths</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;All file paths must be strings, got </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_path</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;File path </span><span class="si">{</span><span class="n">file_path</span><span class="si">}</span><span class="s2"> does not point to an existing file.&quot;</span><span class="p">)</span>

            <span class="c1"># Check file size</span>
            <span class="n">file_size</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">getsize</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">file_size</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">MAX_FILE_SIZE</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;File </span><span class="si">{</span><span class="n">file_path</span><span class="si">}</span><span class="s2"> exceeds maximum size limit of &quot;</span>
                               <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">MAX_FILE_SIZE</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="mi">1024</span><span class="o">*</span><span class="mi">1024</span><span class="p">)</span><span class="si">:</span><span class="s2">.1f</span><span class="si">}</span><span class="s2">MB. &quot;</span>
                               <span class="sa">f</span><span class="s2">&quot;File size: </span><span class="si">{</span><span class="n">file_size</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="mi">1024</span><span class="o">*</span><span class="mi">1024</span><span class="p">)</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">MB&quot;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">file_paths</span> <span class="o">=</span> <span class="n">file_paths</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">size</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">calculate_total_bytes</span><span class="p">()</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">from_folder</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">folder_path</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create Data object from all files in a folder.</span>

<span class="sd">        :param folder_path: Path to directory containing files</span>
<span class="sd">        :return: Data object with all files in the folder</span>
<span class="sd">        :raises TypeError: If folder_path is not a string</span>
<span class="sd">        :raises ValueError: If folder path doesn&#39;t exist or contains no files</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">folder_path</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;folder_path must be a string, got </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">folder_path</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">folder_path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Folder path </span><span class="si">{</span><span class="n">folder_path</span><span class="si">}</span><span class="s2"> does not point to an existing directory.&quot;</span><span class="p">)</span>

        <span class="c1"># Get only files, not directories</span>
        <span class="n">all_items</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">folder_path</span><span class="p">)</span>
        <span class="n">file_paths</span> <span class="o">=</span> <span class="p">[</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">folder_path</span><span class="p">,</span> <span class="n">item</span><span class="p">)</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">all_items</span> 
                     <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">folder_path</span><span class="p">,</span> <span class="n">item</span><span class="p">))]</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">file_paths</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No files found in directory </span><span class="si">{</span><span class="n">folder_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="bp">cls</span><span class="p">(</span><span class="n">file_paths</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">calculate_total_bytes</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate total size of all files in bytes.</span>

<span class="sd">        :return: Total size in bytes</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">total_bytes</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">file_path</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">file_paths</span><span class="p">:</span>
            <span class="n">total_bytes</span> <span class="o">+=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">getsize</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">total_bytes</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        String representation of the Data object.</span>

<span class="sd">        :return: Formatted string with object information</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">output</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;Type: </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">output</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;Total size: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">size</span><span class="si">}</span><span class="s2"> bytes (</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">size</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="mi">1024</span><span class="o">*</span><span class="mi">1024</span><span class="p">)</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> MB)</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">output</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;Number of files: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">file_paths</span><span class="p">)</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">output</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;File paths: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">file_paths</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="k">return</span> <span class="n">output</span>
</code></pre></div></td></tr></table></div>
              </details>



<div class="doc doc-children">










<div class="doc doc-object doc-function">


<h2 id="dnabyte.data_classes.base.Data.__init__" class="doc doc-heading">
            <code class=" language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">file_paths</span><span class="p">)</span></code>

<a href="#dnabyte.data_classes.base.Data.__init__" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Initialize Data object with file paths.</p>
<p>:param file_paths: List of absolute paths to files
:raises TypeError: If file_paths is not a list
:raises ValueError: If file_paths is empty, contains invalid paths, or files exceed size limit</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/data_classes/base.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_paths</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Initialize Data object with file paths.</span>

<span class="sd">    :param file_paths: List of absolute paths to files</span>
<span class="sd">    :raises TypeError: If file_paths is not a list</span>
<span class="sd">    :raises ValueError: If file_paths is empty, contains invalid paths, or files exceed size limit</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Validate input type</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">file_paths</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;file_paths must be a list, got </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">file_paths</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Check if list is empty</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">file_paths</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;file_paths cannot be empty&quot;</span><span class="p">)</span>

    <span class="c1"># Validate each file path</span>
    <span class="k">for</span> <span class="n">file_path</span> <span class="ow">in</span> <span class="n">file_paths</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;All file paths must be strings, got </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;File path </span><span class="si">{</span><span class="n">file_path</span><span class="si">}</span><span class="s2"> does not point to an existing file.&quot;</span><span class="p">)</span>

        <span class="c1"># Check file size</span>
        <span class="n">file_size</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">getsize</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">file_size</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">MAX_FILE_SIZE</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;File </span><span class="si">{</span><span class="n">file_path</span><span class="si">}</span><span class="s2"> exceeds maximum size limit of &quot;</span>
                           <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">MAX_FILE_SIZE</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="mi">1024</span><span class="o">*</span><span class="mi">1024</span><span class="p">)</span><span class="si">:</span><span class="s2">.1f</span><span class="si">}</span><span class="s2">MB. &quot;</span>
                           <span class="sa">f</span><span class="s2">&quot;File size: </span><span class="si">{</span><span class="n">file_size</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="mi">1024</span><span class="o">*</span><span class="mi">1024</span><span class="p">)</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">MB&quot;</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">file_paths</span> <span class="o">=</span> <span class="n">file_paths</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">size</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">calculate_total_bytes</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="dnabyte.data_classes.base.Data.__str__" class="doc doc-heading">
            <code class=" language-python"><span class="fm">__str__</span><span class="p">()</span></code>

<a href="#dnabyte.data_classes.base.Data.__str__" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>String representation of the Data object.</p>
<p>:return: Formatted string with object information</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/data_classes/base.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span>
<span class="normal">95</span>
<span class="normal">96</span>
<span class="normal">97</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    String representation of the Data object.</span>

<span class="sd">    :return: Formatted string with object information</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">output</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;Type: </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="n">output</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;Total size: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">size</span><span class="si">}</span><span class="s2"> bytes (</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">size</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="mi">1024</span><span class="o">*</span><span class="mi">1024</span><span class="p">)</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> MB)</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="n">output</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;Number of files: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">file_paths</span><span class="p">)</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="n">output</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;File paths: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">file_paths</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="k">return</span> <span class="n">output</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="dnabyte.data_classes.base.Data.calculate_total_bytes" class="doc doc-heading">
            <code class=" language-python"><span class="n">calculate_total_bytes</span><span class="p">()</span></code>

<a href="#dnabyte.data_classes.base.Data.calculate_total_bytes" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Calculate total size of all files in bytes.</p>
<p>:return: Total size in bytes</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/data_classes/base.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">calculate_total_bytes</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate total size of all files in bytes.</span>

<span class="sd">    :return: Total size in bytes</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">total_bytes</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">file_path</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">file_paths</span><span class="p">:</span>
        <span class="n">total_bytes</span> <span class="o">+=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">getsize</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">total_bytes</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="dnabyte.data_classes.base.Data.from_folder" class="doc doc-heading">
            <code class=" language-python"><span class="n">from_folder</span><span class="p">(</span><span class="n">folder_path</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

<a href="#dnabyte.data_classes.base.Data.from_folder" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Create Data object from all files in a folder.</p>
<p>:param folder_path: Path to directory containing files
:return: Data object with all files in the folder
:raises TypeError: If folder_path is not a string
:raises ValueError: If folder path doesn't exist or contains no files</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/data_classes/base.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">from_folder</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">folder_path</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create Data object from all files in a folder.</span>

<span class="sd">    :param folder_path: Path to directory containing files</span>
<span class="sd">    :return: Data object with all files in the folder</span>
<span class="sd">    :raises TypeError: If folder_path is not a string</span>
<span class="sd">    :raises ValueError: If folder path doesn&#39;t exist or contains no files</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">folder_path</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;folder_path must be a string, got </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">folder_path</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">folder_path</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Folder path </span><span class="si">{</span><span class="n">folder_path</span><span class="si">}</span><span class="s2"> does not point to an existing directory.&quot;</span><span class="p">)</span>

    <span class="c1"># Get only files, not directories</span>
    <span class="n">all_items</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">folder_path</span><span class="p">)</span>
    <span class="n">file_paths</span> <span class="o">=</span> <span class="p">[</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">folder_path</span><span class="p">,</span> <span class="n">item</span><span class="p">)</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">all_items</span> 
                 <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">folder_path</span><span class="p">,</span> <span class="n">item</span><span class="p">))]</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">file_paths</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No files found in directory </span><span class="si">{</span><span class="n">folder_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="bp">cls</span><span class="p">(</span><span class="n">file_paths</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div><h3 id="binarycode">BinaryCode<a class="headerlink" href="#binarycode" title="Permanent link">&para;</a></h3>


<div class="doc doc-object doc-class">



<a id="dnabyte.data_classes.binarycode.BinaryCode"></a>
    <div class="doc doc-contents first">
            <p class="doc doc-class-bases">
              Bases: <code><a class="autorefs autorefs-internal" title="dnabyte.data_classes.base.Data" href="#dnabyte.data_classes.base.Data">Data</a></code></p>



        <p>Represents binary data as a sequence of '0' and '1' characters.</p>
<p>BinaryCode stores and manipulates binary data streams created from file 
binarization processes. It provides validation, indexing, and iteration 
capabilities for working with binary representations of original data.</p>
<p>Extends the Data base class to maintain compatibility with the DNAbyte 
pipeline while providing binary-specific functionality.</p>


<details class="can-be-created-either" open>
  <summary>Can be created either</summary>
  <ul>
<li>from file data using a Binarize method, or</li>
<li>directly by providing a binary string.</li>
</ul>
</details>        <p>:param bitstream: A string of binary data (only '0' and '1' characters).
:param file_paths: Optional list of original file paths (for inheritance compatibility).
:param size: Optional original file size in bytes.
:raises TypeError: If bitstream is not a string.
:raises ValueError: If bitstream is empty or contains non-binary characters.</p>


<details class="example" open>
  <summary>Example</summary>
  <blockquote>
<blockquote>
<blockquote>
<p>binary = BinaryCode("10110010", file_paths=["/path/to/file.txt"])
print(len(binary))  # 8
print(binary[0])    # '1'</p>
</blockquote>
</blockquote>
</blockquote>
</details>







              <details class="mkdocstrings-source">
                <summary>Source code in <code>dnabyte/data_classes/binarycode.py</code></summary>
                <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">  4</span>
<span class="normal">  5</span>
<span class="normal">  6</span>
<span class="normal">  7</span>
<span class="normal">  8</span>
<span class="normal">  9</span>
<span class="normal"> 10</span>
<span class="normal"> 11</span>
<span class="normal"> 12</span>
<span class="normal"> 13</span>
<span class="normal"> 14</span>
<span class="normal"> 15</span>
<span class="normal"> 16</span>
<span class="normal"> 17</span>
<span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">BinaryCode</span><span class="p">(</span><span class="n">Data</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Represents binary data as a sequence of &#39;0&#39; and &#39;1&#39; characters.</span>

<span class="sd">    BinaryCode stores and manipulates binary data streams created from file </span>
<span class="sd">    binarization processes. It provides validation, indexing, and iteration </span>
<span class="sd">    capabilities for working with binary representations of original data.</span>

<span class="sd">    Extends the Data base class to maintain compatibility with the DNAbyte </span>
<span class="sd">    pipeline while providing binary-specific functionality.</span>

<span class="sd">    Can be created either:</span>
<span class="sd">        - from file data using a Binarize method, or</span>
<span class="sd">        - directly by providing a binary string.</span>

<span class="sd">    :param bitstream: A string of binary data (only &#39;0&#39; and &#39;1&#39; characters).</span>
<span class="sd">    :param file_paths: Optional list of original file paths (for inheritance compatibility).</span>
<span class="sd">    :param size: Optional original file size in bytes.</span>
<span class="sd">    :raises TypeError: If bitstream is not a string.</span>
<span class="sd">    :raises ValueError: If bitstream is empty or contains non-binary characters.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; binary = BinaryCode(&quot;10110010&quot;, file_paths=[&quot;/path/to/file.txt&quot;])</span>
<span class="sd">        &gt;&gt;&gt; print(len(binary))  # 8</span>
<span class="sd">        &gt;&gt;&gt; print(binary[0])    # &#39;1&#39;</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">file_paths</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Creates a BinaryCode object from a bitstream.</span>

<span class="sd">        :param bitstream: A string of binary data.</span>
<span class="sd">        :param file_paths: Optional list of original file paths.</span>
<span class="sd">        :param size: Optional original file size in bytes.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Validate the bitstream before setting any attributes</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_bitstream</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

        <span class="c1"># Set binary-specific attributes</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

        <span class="c1"># Handle inheritance - set attributes for parent class compatibility</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">file_paths</span> <span class="o">=</span> <span class="n">file_paths</span> <span class="ow">or</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">size</span> <span class="o">=</span> <span class="n">size</span>  <span class="c1"># Keep track of original file size</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_validate_bitstream</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Validates that the input is a proper bitstream.</span>

<span class="sd">        :param bitstream: The data to validate</span>
<span class="sd">        :raises TypeError: If bitstream is not a string</span>
<span class="sd">        :raises ValueError: If bitstream is empty or contains invalid characters</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Check if data is a string</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Bitstream must be a string, got </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">data</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>

        <span class="c1"># Check if data is empty</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Bitstream cannot be empty&quot;</span><span class="p">)</span>

        <span class="c1"># Check if all characters are either &#39;0&#39; or &#39;1&#39;</span>
        <span class="n">invalid_chars</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">char</span> <span class="k">for</span> <span class="n">char</span> <span class="ow">in</span> <span class="n">data</span> <span class="k">if</span> <span class="n">char</span> <span class="ow">not</span> <span class="ow">in</span> <span class="s2">&quot;01&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">invalid_chars</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Bitstream contains invalid characters: </span><span class="si">{</span><span class="n">invalid_chars</span><span class="si">}</span><span class="s2">. &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;Only &#39;0&#39; and &#39;1&#39; are allowed.&quot;</span>
            <span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">compare_binary_strings</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">bin_str1</span><span class="p">,</span> <span class="n">bin_str2</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Helper function for compare(). Compare two binary strings and return the indices of the differing bits.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">differences</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">min_length</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">bin_str1</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">bin_str2</span><span class="p">))</span>

        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">min_length</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">bin_str1</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">!=</span> <span class="n">bin_str2</span><span class="p">[</span><span class="n">i</span><span class="p">]:</span>
                <span class="n">differences</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">differences</span>


    <span class="k">def</span><span class="w"> </span><span class="nf">compare</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data_dec</span><span class="p">,</span> <span class="n">data_bin</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Compare decoded data with raw data and return the result of the comparison.</span>

<span class="sd">        Args:</span>
<span class="sd">            data_dec (BinaryCode): The decoded data object.</span>
<span class="sd">            data_bin (BinaryCode): The raw data object.</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: A message indicating whether the data was successfully decoded or not.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">res</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">compare_binary_strings</span><span class="p">(</span><span class="n">data_dec</span><span class="o">.</span><span class="n">data</span><span class="p">,</span> <span class="n">data_bin</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">res</span> <span class="o">==</span> <span class="p">[]</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">data_dec</span><span class="o">.</span><span class="n">data</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">data_bin</span><span class="o">.</span><span class="n">data</span><span class="p">):</span>
            <span class="k">return</span> <span class="s1">&#39;SUCCESS&#39;</span><span class="p">,</span> <span class="n">res</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="s1">&#39;ERROR&#39;</span><span class="p">,</span> <span class="n">res</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">random</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generates a random BinaryCode object with a specified length.</span>

<span class="sd">        :param n: Length of the binary string to generate.</span>
<span class="sd">        :return: BinaryCode object with random binary data.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">bitstream</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="s1">&#39;01&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">BinaryCode</span><span class="p">(</span><span class="n">bitstream</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">output</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;Type: </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">output</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;Length: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="si">}</span><span class="s2"> bits</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">size</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">output</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;Size: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">size</span><span class="si">}</span><span class="s2"> bytes</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">output</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;Data: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">50</span><span class="p">]</span><span class="si">}</span><span class="s2">...</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="k">return</span> <span class="n">output</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Developer-friendly representation of the BinaryCode object.</span>

<span class="sd">        :return: Unambiguous string representation</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Truncate data if too long for readable repr</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">20</span><span class="p">:</span>
            <span class="n">data_repr</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[:</span><span class="mi">20</span><span class="p">]</span><span class="si">}</span><span class="s2">...&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">data_repr</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span>

        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;BinaryCode(bitstream=&#39;</span><span class="si">{</span><span class="n">data_repr</span><span class="si">}</span><span class="s2">&#39;, length=</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="si">}</span><span class="s2">)&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return the number of bits in the bitstream.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Allow indexing to get individual bits or slices.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="n">index</span><span class="p">]</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Allow iteration over individual bits.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">iter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



<div class="doc doc-children">










<div class="doc doc-object doc-function">


<h2 id="dnabyte.data_classes.binarycode.BinaryCode.__getitem__" class="doc doc-heading">
            <code class=" language-python"><span class="fm">__getitem__</span><span class="p">(</span><span class="n">index</span><span class="p">)</span></code>

<a href="#dnabyte.data_classes.binarycode.BinaryCode.__getitem__" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Allow indexing to get individual bits or slices.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/data_classes/binarycode.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Allow indexing to get individual bits or slices.&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="n">index</span><span class="p">]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="dnabyte.data_classes.binarycode.BinaryCode.__init__" class="doc doc-heading">
            <code class=" language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">file_paths</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#dnabyte.data_classes.binarycode.BinaryCode.__init__" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Creates a BinaryCode object from a bitstream.</p>
<p>:param bitstream: A string of binary data.
:param file_paths: Optional list of original file paths.
:param size: Optional original file size in bytes.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/data_classes/binarycode.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">file_paths</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates a BinaryCode object from a bitstream.</span>

<span class="sd">    :param bitstream: A string of binary data.</span>
<span class="sd">    :param file_paths: Optional list of original file paths.</span>
<span class="sd">    :param size: Optional original file size in bytes.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Validate the bitstream before setting any attributes</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_validate_bitstream</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

    <span class="c1"># Set binary-specific attributes</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

    <span class="c1"># Handle inheritance - set attributes for parent class compatibility</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">file_paths</span> <span class="o">=</span> <span class="n">file_paths</span> <span class="ow">or</span> <span class="p">[]</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">size</span> <span class="o">=</span> <span class="n">size</span>  <span class="c1"># Keep track of original file size</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="dnabyte.data_classes.binarycode.BinaryCode.__iter__" class="doc doc-heading">
            <code class=" language-python"><span class="fm">__iter__</span><span class="p">()</span></code>

<a href="#dnabyte.data_classes.binarycode.BinaryCode.__iter__" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Allow iteration over individual bits.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/data_classes/binarycode.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="fm">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Allow iteration over individual bits.&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">iter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="dnabyte.data_classes.binarycode.BinaryCode.__len__" class="doc doc-heading">
            <code class=" language-python"><span class="fm">__len__</span><span class="p">()</span></code>

<a href="#dnabyte.data_classes.binarycode.BinaryCode.__len__" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Return the number of bits in the bitstream.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/data_classes/binarycode.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Return the number of bits in the bitstream.&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="dnabyte.data_classes.binarycode.BinaryCode.__repr__" class="doc doc-heading">
            <code class=" language-python"><span class="fm">__repr__</span><span class="p">()</span></code>

<a href="#dnabyte.data_classes.binarycode.BinaryCode.__repr__" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Developer-friendly representation of the BinaryCode object.</p>
<p>:return: Unambiguous string representation</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/data_classes/binarycode.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Developer-friendly representation of the BinaryCode object.</span>

<span class="sd">    :return: Unambiguous string representation</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Truncate data if too long for readable repr</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">20</span><span class="p">:</span>
        <span class="n">data_repr</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[:</span><span class="mi">20</span><span class="p">]</span><span class="si">}</span><span class="s2">...&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">data_repr</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span>

    <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;BinaryCode(bitstream=&#39;</span><span class="si">{</span><span class="n">data_repr</span><span class="si">}</span><span class="s2">&#39;, length=</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">length</span><span class="si">}</span><span class="s2">)&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="dnabyte.data_classes.binarycode.BinaryCode.compare" class="doc doc-heading">
            <code class=" language-python"><span class="n">compare</span><span class="p">(</span><span class="n">data_dec</span><span class="p">,</span> <span class="n">data_bin</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#dnabyte.data_classes.binarycode.BinaryCode.compare" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Compare decoded data with raw data and return the result of the comparison.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>data_dec</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="dnabyte.data_classes.binarycode.BinaryCode" href="#dnabyte.data_classes.binarycode.BinaryCode">BinaryCode</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The decoded data object.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>data_bin</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="dnabyte.data_classes.binarycode.BinaryCode" href="#dnabyte.data_classes.binarycode.BinaryCode">BinaryCode</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The raw data object.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A message indicating whether the data was successfully decoded or not.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/data_classes/binarycode.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">compare</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data_dec</span><span class="p">,</span> <span class="n">data_bin</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Compare decoded data with raw data and return the result of the comparison.</span>

<span class="sd">    Args:</span>
<span class="sd">        data_dec (BinaryCode): The decoded data object.</span>
<span class="sd">        data_bin (BinaryCode): The raw data object.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: A message indicating whether the data was successfully decoded or not.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">res</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">compare_binary_strings</span><span class="p">(</span><span class="n">data_dec</span><span class="o">.</span><span class="n">data</span><span class="p">,</span> <span class="n">data_bin</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">res</span> <span class="o">==</span> <span class="p">[]</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">data_dec</span><span class="o">.</span><span class="n">data</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">data_bin</span><span class="o">.</span><span class="n">data</span><span class="p">):</span>
        <span class="k">return</span> <span class="s1">&#39;SUCCESS&#39;</span><span class="p">,</span> <span class="n">res</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="s1">&#39;ERROR&#39;</span><span class="p">,</span> <span class="n">res</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="dnabyte.data_classes.binarycode.BinaryCode.compare_binary_strings" class="doc doc-heading">
            <code class=" language-python"><span class="n">compare_binary_strings</span><span class="p">(</span><span class="n">bin_str1</span><span class="p">,</span> <span class="n">bin_str2</span><span class="p">)</span></code>

<a href="#dnabyte.data_classes.binarycode.BinaryCode.compare_binary_strings" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Helper function for compare(). Compare two binary strings and return the indices of the differing bits.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/data_classes/binarycode.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">compare_binary_strings</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">bin_str1</span><span class="p">,</span> <span class="n">bin_str2</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Helper function for compare(). Compare two binary strings and return the indices of the differing bits.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">differences</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">min_length</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">bin_str1</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">bin_str2</span><span class="p">))</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">min_length</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">bin_str1</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">!=</span> <span class="n">bin_str2</span><span class="p">[</span><span class="n">i</span><span class="p">]:</span>
            <span class="n">differences</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">differences</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="dnabyte.data_classes.binarycode.BinaryCode.random" class="doc doc-heading">
            <code class=" language-python"><span class="n">random</span><span class="p">(</span><span class="n">n</span><span class="p">)</span></code>

<a href="#dnabyte.data_classes.binarycode.BinaryCode.random" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Generates a random BinaryCode object with a specified length.</p>
<p>:param n: Length of the binary string to generate.
:return: BinaryCode object with random binary data.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/data_classes/binarycode.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">random</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generates a random BinaryCode object with a specified length.</span>

<span class="sd">    :param n: Length of the binary string to generate.</span>
<span class="sd">    :return: BinaryCode object with random binary data.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">bitstream</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="s1">&#39;01&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">BinaryCode</span><span class="p">(</span><span class="n">bitstream</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div><h3 id="nucleobasecode">NucleobaseCode<a class="headerlink" href="#nucleobasecode" title="Permanent link">&para;</a></h3>


<div class="doc doc-object doc-class">



<a id="dnabyte.data_classes.nucleobasecode.NucleobaseCode"></a>
    <div class="doc doc-contents first">
            <p class="doc doc-class-bases">
              Bases: <code><a class="autorefs autorefs-internal" title="dnabyte.data_classes.base.Data" href="#dnabyte.data_classes.base.Data">Data</a></code></p>



        <p>NucleobaseCode is a subclass of Data. Its main attribute is the data object, which consists of a multi-leveled list. At the highest level, data consists of a list of codewords, there is one list for every codeword in the data object.
At the lowest level there are lists of motifs or nucleotides, that are to be hybridised in a single pool. The list
structure represents the hierarchical structure of the pooling process.</p>


<details class="it-can-be-instantiated-either-by" open>
  <summary>It can be instantiated either by</summary>
  <ul>
<li>by an encoder producing the encoded data structure or</li>
<li>directly providing the encoded data structure as a parameter.</li>
</ul>
</details>        <p>:param data: The encoded data structure (multi-level list of codewords).
:param file_paths: Optional list of original file paths.
:param size: Optional size information from original data.
:raises TypeError: If data is not a list.
:raises ValueError: If data is empty or has invalid structure.</p>








              <details class="mkdocstrings-source">
                <summary>Source code in <code>dnabyte/data_classes/nucleobasecode.py</code></summary>
                <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">  5</span>
<span class="normal">  6</span>
<span class="normal">  7</span>
<span class="normal">  8</span>
<span class="normal">  9</span>
<span class="normal"> 10</span>
<span class="normal"> 11</span>
<span class="normal"> 12</span>
<span class="normal"> 13</span>
<span class="normal"> 14</span>
<span class="normal"> 15</span>
<span class="normal"> 16</span>
<span class="normal"> 17</span>
<span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">NucleobaseCode</span><span class="p">(</span><span class="n">Data</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    NucleobaseCode is a subclass of Data. Its main attribute is the data object, which consists of a multi-leveled list. At the highest level, data consists of a list of codewords, there is one list for every codeword in the data object.</span>
<span class="sd">    At the lowest level there are lists of motifs or nucleotides, that are to be hybridised in a single pool. The list</span>
<span class="sd">    structure represents the hierarchical structure of the pooling process.</span>

<span class="sd">    It can be instantiated either by:</span>
<span class="sd">        - by an encoder producing the encoded data structure or</span>
<span class="sd">        - directly providing the encoded data structure as a parameter.</span>

<span class="sd">    :param data: The encoded data structure (multi-level list of codewords).</span>
<span class="sd">    :param file_paths: Optional list of original file paths.</span>
<span class="sd">    :param size: Optional size information from original data.</span>
<span class="sd">    :raises TypeError: If data is not a list.</span>
<span class="sd">    :raises ValueError: If data is empty or has invalid structure.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Creates a NucleobaseCode object from encoded data structure.</span>

<span class="sd">        :param data: Multi-level list structure representing encoded codewords.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Validate the encoded data before setting any attributes</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_data</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">num_codewords</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)</span> <span class="k">if</span> <span class="n">data</span> <span class="k">else</span> <span class="mi">0</span>

        <span class="c1"># Calculate additional metrics</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">max_depth</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_calculate_max_depth</span><span class="p">()</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_validate_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Validates that the input is a proper encoded data structure.</span>

<span class="sd">        :param data: The data to validate</span>
<span class="sd">        :raises TypeError: If data is not a list</span>
<span class="sd">        :raises ValueError: If data is empty or has invalid structure</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Check if data is a list</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Encoded data must be a list, got </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">data</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>

        <span class="c1"># Check if data is empty</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Encoded data cannot be empty&quot;</span><span class="p">)</span>

        <span class="c1"># TODO: Check structure of the codewords</span>
        <span class="c1"># TODO: Check whether each codeword contains only valid nucleotides</span>
        <span class="c1"># self._validate_codeword_structure(data)</span>


    <span class="c1"># TODO: calculate_max_depth delivers incorrect results in some tests</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_calculate_max_depth</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculates the maximum depth of the nested structure.</span>

<span class="sd">        :return: Maximum nesting depth</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">:</span>
            <span class="k">return</span> <span class="mi">0</span>

        <span class="n">max_depth</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">codeword</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">:</span>
            <span class="n">depth</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_depth_recursive</span><span class="p">(</span><span class="n">codeword</span><span class="p">)</span>
            <span class="n">max_depth</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">max_depth</span><span class="p">,</span> <span class="n">depth</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">max_depth</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_get_depth_recursive</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">structure</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Recursively calculates depth of nested structure.</span>

<span class="sd">        :param structure: The structure to measure depth of</span>
<span class="sd">        :return: Depth of structure</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">structure</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">structure</span><span class="p">:</span>
                <span class="k">return</span> <span class="mi">1</span>
            <span class="k">return</span> <span class="mi">1</span> <span class="o">+</span> <span class="nb">max</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_get_depth_recursive</span><span class="p">(</span><span class="n">item</span><span class="p">)</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">structure</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="mi">0</span>  <span class="c1"># Nucleotide has no depth</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">get_codeword</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get a specific codeword by index.</span>

<span class="sd">        :param index: Index of the codeword</span>
<span class="sd">        :return: The codeword at the specified index</span>
<span class="sd">        :raises IndexError: If index is out of range</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="mi">0</span> <span class="o">&lt;=</span> <span class="n">index</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">IndexError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Codeword index </span><span class="si">{</span><span class="n">index</span><span class="si">}</span><span class="s2"> out of range (0-</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="si">}</span><span class="s2">)&quot;</span>
            <span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="n">index</span><span class="p">]</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">validate</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Re-validates the current encoded data structure.</span>

<span class="sd">        :return: True if valid</span>
<span class="sd">        :raises: Various exceptions if invalid</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_data</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">True</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">random</span><span class="p">(</span><span class="nb">type</span><span class="o">=</span><span class="s1">&#39;synthesis&#39;</span><span class="p">,</span> <span class="n">library</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">m</span><span class="o">=</span><span class="mi">250</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generates a random NucleobaseCode object with a specified number of codewords.</span>

<span class="sd">        :param type: Type of nucleobase code (e.g., &#39;DNA&#39;, &#39;RNA&#39;)</span>
<span class="sd">        :param n: Number of codewords to generate</span>
<span class="sd">        :return: NucleobaseCode object with random data</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">data</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">if</span> <span class="nb">type</span> <span class="o">==</span> <span class="s1">&#39;synthesis&#39;</span><span class="p">:</span>

            <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
                <span class="n">codeword</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">([</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">])</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">m</span><span class="p">))</span>
                <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">codeword</span><span class="p">)</span>

        <span class="k">elif</span> <span class="nb">type</span> <span class="o">==</span> <span class="s1">&#39;nested_assembly&#39;</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">library</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Library must be provided for assembly type&quot;</span><span class="p">)</span>

            <span class="k">else</span><span class="p">:</span>
                <span class="n">library</span> <span class="o">=</span> <span class="n">Library</span><span class="p">(</span><span class="n">structure</span><span class="o">=</span><span class="s1">&#39;linear_assembly&#39;</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="s1">&#39;./tests/testlibraries/20bp_Lib.csv&#39;</span><span class="p">)</span>

                <span class="c1"># Helper function to create nested list of random depth</span>
                <span class="k">def</span><span class="w"> </span><span class="nf">create_nested_structure</span><span class="p">(</span><span class="n">current_depth</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">):</span>
<span class="w">                    </span><span class="sd">&quot;&quot;&quot;Recursively create nested list structure with random depth.&quot;&quot;&quot;</span>
                    <span class="c1"># If we&#39;ve reached max depth or randomly decide to stop, return a leaf element</span>
                    <span class="k">if</span> <span class="n">current_depth</span> <span class="o">&gt;=</span> <span class="n">max_depth</span> <span class="ow">or</span> <span class="p">(</span><span class="n">current_depth</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">()</span> <span class="o">&lt;</span> <span class="mf">0.3</span><span class="p">):</span>
                        <span class="k">return</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">library</span><span class="o">.</span><span class="n">library</span><span class="p">)</span>

                    <span class="c1"># Create a list with 1-3 children</span>
                    <span class="n">num_children</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
                    <span class="k">return</span> <span class="p">[</span><span class="n">create_nested_structure</span><span class="p">(</span><span class="n">current_depth</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">)</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_children</span><span class="p">)]</span>

                <span class="c1"># Generate n codewords, each with random nested structure (depth 1-5)</span>
                <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
                    <span class="n">max_depth</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">5</span><span class="p">)</span>
                    <span class="n">codeword</span> <span class="o">=</span> <span class="n">create_nested_structure</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">)</span>
                    <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">codeword</span><span class="p">)</span>


        <span class="k">elif</span> <span class="nb">type</span> <span class="o">==</span> <span class="s1">&#39;linear_assembly&#39;</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">library</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Library must be provided for assembly type&quot;</span><span class="p">)</span>

            <span class="k">else</span><span class="p">:</span>
                <span class="n">library</span> <span class="o">=</span> <span class="n">Library</span><span class="p">(</span><span class="n">structure</span><span class="o">=</span><span class="s1">&#39;linear_assembly&#39;</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="s1">&#39;./tests/testlibraries/20bp_Lib.csv&#39;</span><span class="p">)</span>

                <span class="c1"># Generate n codewords, each with m library elements</span>
                <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
                    <span class="n">codeword</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">choices</span><span class="p">(</span><span class="n">library</span><span class="o">.</span><span class="n">library</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="n">m</span><span class="p">)</span>
                    <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">codeword</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid synthesis type: </span><span class="si">{</span><span class="nb">type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">NucleobaseCode</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>


    <span class="k">def</span><span class="w"> </span><span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">output</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;Type: </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">output</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;Number of codewords: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">num_codewords</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">output</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;Max depth: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">max_depth</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>

        <span class="c1"># Add file paths if they exist</span>
        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s2">&quot;file_paths&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">file_paths</span><span class="p">:</span>
            <span class="n">output</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;File paths: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">file_paths</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>

        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s2">&quot;size&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">size</span><span class="p">:</span>
            <span class="n">output</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;Original size: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">size</span><span class="si">}</span><span class="s2"> bytes</span><span class="se">\n</span><span class="s2">&quot;</span>

        <span class="n">output</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;DATA: </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)[:</span><span class="mi">100</span><span class="p">]</span><span class="si">}</span><span class="s2">...</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="k">return</span> <span class="n">output</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return the number of codewords.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Allow indexing to get codewords.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_codeword</span><span class="p">(</span><span class="n">index</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



<div class="doc doc-children">










<div class="doc doc-object doc-function">


<h2 id="dnabyte.data_classes.nucleobasecode.NucleobaseCode.__getitem__" class="doc doc-heading">
            <code class=" language-python"><span class="fm">__getitem__</span><span class="p">(</span><span class="n">index</span><span class="p">)</span></code>

<a href="#dnabyte.data_classes.nucleobasecode.NucleobaseCode.__getitem__" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Allow indexing to get codewords.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/data_classes/nucleobasecode.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Allow indexing to get codewords.&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_codeword</span><span class="p">(</span><span class="n">index</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="dnabyte.data_classes.nucleobasecode.NucleobaseCode.__init__" class="doc doc-heading">
            <code class=" language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></code>

<a href="#dnabyte.data_classes.nucleobasecode.NucleobaseCode.__init__" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Creates a NucleobaseCode object from encoded data structure.</p>
<p>:param data: Multi-level list structure representing encoded codewords.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/data_classes/nucleobasecode.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates a NucleobaseCode object from encoded data structure.</span>

<span class="sd">    :param data: Multi-level list structure representing encoded codewords.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Validate the encoded data before setting any attributes</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_validate_data</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">num_codewords</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)</span> <span class="k">if</span> <span class="n">data</span> <span class="k">else</span> <span class="mi">0</span>

    <span class="c1"># Calculate additional metrics</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">max_depth</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_calculate_max_depth</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="dnabyte.data_classes.nucleobasecode.NucleobaseCode.__len__" class="doc doc-heading">
            <code class=" language-python"><span class="fm">__len__</span><span class="p">()</span></code>

<a href="#dnabyte.data_classes.nucleobasecode.NucleobaseCode.__len__" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Return the number of codewords.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/data_classes/nucleobasecode.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Return the number of codewords.&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="dnabyte.data_classes.nucleobasecode.NucleobaseCode.get_codeword" class="doc doc-heading">
            <code class=" language-python"><span class="n">get_codeword</span><span class="p">(</span><span class="n">index</span><span class="p">)</span></code>

<a href="#dnabyte.data_classes.nucleobasecode.NucleobaseCode.get_codeword" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Get a specific codeword by index.</p>
<p>:param index: Index of the codeword
:return: The codeword at the specified index
:raises IndexError: If index is out of range</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/data_classes/nucleobasecode.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_codeword</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get a specific codeword by index.</span>

<span class="sd">    :param index: Index of the codeword</span>
<span class="sd">    :return: The codeword at the specified index</span>
<span class="sd">    :raises IndexError: If index is out of range</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="mi">0</span> <span class="o">&lt;=</span> <span class="n">index</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">IndexError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Codeword index </span><span class="si">{</span><span class="n">index</span><span class="si">}</span><span class="s2"> out of range (0-</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="si">}</span><span class="s2">)&quot;</span>
        <span class="p">)</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="n">index</span><span class="p">]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="dnabyte.data_classes.nucleobasecode.NucleobaseCode.random" class="doc doc-heading">
            <code class=" language-python"><span class="n">random</span><span class="p">(</span><span class="nb">type</span><span class="o">=</span><span class="s1">&#39;synthesis&#39;</span><span class="p">,</span> <span class="n">library</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">m</span><span class="o">=</span><span class="mi">250</span><span class="p">)</span></code>

<a href="#dnabyte.data_classes.nucleobasecode.NucleobaseCode.random" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Generates a random NucleobaseCode object with a specified number of codewords.</p>
<p>:param type: Type of nucleobase code (e.g., 'DNA', 'RNA')
:param n: Number of codewords to generate
:return: NucleobaseCode object with random data</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/data_classes/nucleobasecode.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">random</span><span class="p">(</span><span class="nb">type</span><span class="o">=</span><span class="s1">&#39;synthesis&#39;</span><span class="p">,</span> <span class="n">library</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">m</span><span class="o">=</span><span class="mi">250</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generates a random NucleobaseCode object with a specified number of codewords.</span>

<span class="sd">    :param type: Type of nucleobase code (e.g., &#39;DNA&#39;, &#39;RNA&#39;)</span>
<span class="sd">    :param n: Number of codewords to generate</span>
<span class="sd">    :return: NucleobaseCode object with random data</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">data</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">if</span> <span class="nb">type</span> <span class="o">==</span> <span class="s1">&#39;synthesis&#39;</span><span class="p">:</span>

        <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
            <span class="n">codeword</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">([</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">])</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">m</span><span class="p">))</span>
            <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">codeword</span><span class="p">)</span>

    <span class="k">elif</span> <span class="nb">type</span> <span class="o">==</span> <span class="s1">&#39;nested_assembly&#39;</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">library</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Library must be provided for assembly type&quot;</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="n">library</span> <span class="o">=</span> <span class="n">Library</span><span class="p">(</span><span class="n">structure</span><span class="o">=</span><span class="s1">&#39;linear_assembly&#39;</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="s1">&#39;./tests/testlibraries/20bp_Lib.csv&#39;</span><span class="p">)</span>

            <span class="c1"># Helper function to create nested list of random depth</span>
            <span class="k">def</span><span class="w"> </span><span class="nf">create_nested_structure</span><span class="p">(</span><span class="n">current_depth</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">):</span>
<span class="w">                </span><span class="sd">&quot;&quot;&quot;Recursively create nested list structure with random depth.&quot;&quot;&quot;</span>
                <span class="c1"># If we&#39;ve reached max depth or randomly decide to stop, return a leaf element</span>
                <span class="k">if</span> <span class="n">current_depth</span> <span class="o">&gt;=</span> <span class="n">max_depth</span> <span class="ow">or</span> <span class="p">(</span><span class="n">current_depth</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">()</span> <span class="o">&lt;</span> <span class="mf">0.3</span><span class="p">):</span>
                    <span class="k">return</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">library</span><span class="o">.</span><span class="n">library</span><span class="p">)</span>

                <span class="c1"># Create a list with 1-3 children</span>
                <span class="n">num_children</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
                <span class="k">return</span> <span class="p">[</span><span class="n">create_nested_structure</span><span class="p">(</span><span class="n">current_depth</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">)</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_children</span><span class="p">)]</span>

            <span class="c1"># Generate n codewords, each with random nested structure (depth 1-5)</span>
            <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
                <span class="n">max_depth</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">5</span><span class="p">)</span>
                <span class="n">codeword</span> <span class="o">=</span> <span class="n">create_nested_structure</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">)</span>
                <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">codeword</span><span class="p">)</span>


    <span class="k">elif</span> <span class="nb">type</span> <span class="o">==</span> <span class="s1">&#39;linear_assembly&#39;</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">library</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Library must be provided for assembly type&quot;</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="n">library</span> <span class="o">=</span> <span class="n">Library</span><span class="p">(</span><span class="n">structure</span><span class="o">=</span><span class="s1">&#39;linear_assembly&#39;</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="s1">&#39;./tests/testlibraries/20bp_Lib.csv&#39;</span><span class="p">)</span>

            <span class="c1"># Generate n codewords, each with m library elements</span>
            <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
                <span class="n">codeword</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">choices</span><span class="p">(</span><span class="n">library</span><span class="o">.</span><span class="n">library</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="n">m</span><span class="p">)</span>
                <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">codeword</span><span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid synthesis type: </span><span class="si">{</span><span class="nb">type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">NucleobaseCode</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="dnabyte.data_classes.nucleobasecode.NucleobaseCode.validate" class="doc doc-heading">
            <code class=" language-python"><span class="n">validate</span><span class="p">()</span></code>

<a href="#dnabyte.data_classes.nucleobasecode.NucleobaseCode.validate" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Re-validates the current encoded data structure.</p>
<p>:return: True if valid
:raises: Various exceptions if invalid</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/data_classes/nucleobasecode.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">validate</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Re-validates the current encoded data structure.</span>

<span class="sd">    :return: True if valid</span>
<span class="sd">    :raises: Various exceptions if invalid</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_validate_data</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
    <span class="k">return</span> <span class="kc">True</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div><h3 id="insilicodna">InSilicoDNA<a class="headerlink" href="#insilicodna" title="Permanent link">&para;</a></h3>


<div class="doc doc-object doc-class">



<a id="dnabyte.data_classes.insilicodna.InSilicoDNA"></a>
    <div class="doc doc-contents first">
            <p class="doc doc-class-bases">
              Bases: <code><a class="autorefs autorefs-internal" title="dnabyte.data_classes.base.Data" href="#dnabyte.data_classes.base.Data">Data</a></code></p>



        <p>Represents synthesized DNA sequences as oligonucleotide strings.</p>
<p>InSilicoDNA contains a collection of DNA sequences (strings of 'A', 'C', 'G', 'T')
that represent synthesized DNA. This class is used for DNA synthesis simulation
and applying error channels like storage simulation and PCR amplification.</p>
<p>Inherits from Data to maintain compatibility with the DNAbyte pipeline.</p>


<details class="it-can-be-instantiated-either" open>
  <summary>It can be instantiated either</summary>
  <ul>
<li>by a synthesis simulator producing the synthesised DNA sequences, or</li>
<li>directly providing the list of DNA sequences as a parameter.</li>
</ul>
</details>        <p>:param data: List of DNA sequences (strings of nucleotides).
:raises TypeError: If data is not a list.
:raises ValueError: If data is empty or contains invalid DNA sequences.</p>








              <details class="mkdocstrings-source">
                <summary>Source code in <code>dnabyte/data_classes/insilicodna.py</code></summary>
                <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">  4</span>
<span class="normal">  5</span>
<span class="normal">  6</span>
<span class="normal">  7</span>
<span class="normal">  8</span>
<span class="normal">  9</span>
<span class="normal"> 10</span>
<span class="normal"> 11</span>
<span class="normal"> 12</span>
<span class="normal"> 13</span>
<span class="normal"> 14</span>
<span class="normal"> 15</span>
<span class="normal"> 16</span>
<span class="normal"> 17</span>
<span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">InSilicoDNA</span><span class="p">(</span><span class="n">Data</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Represents synthesized DNA sequences as oligonucleotide strings.</span>

<span class="sd">    InSilicoDNA contains a collection of DNA sequences (strings of &#39;A&#39;, &#39;C&#39;, &#39;G&#39;, &#39;T&#39;)</span>
<span class="sd">    that represent synthesized DNA. This class is used for DNA synthesis simulation</span>
<span class="sd">    and applying error channels like storage simulation and PCR amplification.</span>

<span class="sd">    Inherits from Data to maintain compatibility with the DNAbyte pipeline.</span>

<span class="sd">    It can be instantiated either:</span>
<span class="sd">        - by a synthesis simulator producing the synthesised DNA sequences, or</span>
<span class="sd">        - directly providing the list of DNA sequences as a parameter.</span>

<span class="sd">    :param data: List of DNA sequences (strings of nucleotides).</span>
<span class="sd">    :raises TypeError: If data is not a list.</span>
<span class="sd">    :raises ValueError: If data is empty or contains invalid DNA sequences.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Creates a DNA object from DNA sequences.</span>

<span class="sd">        :param data: List of DNA sequences (strings).</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Initialize directly from DNA sequences</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_dna_data</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">file_paths</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">size</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="c1"># Calculate metrics</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">num_sequences</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="k">else</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">total_length</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="k">else</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">average_length</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">total_length</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_sequences</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_sequences</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_validate_dna_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Validates that the input is a proper list of DNA sequences.</span>

<span class="sd">        :param data: The data to validate</span>
<span class="sd">        :raises TypeError: If data is not a list</span>
<span class="sd">        :raises ValueError: If data is empty or contains invalid sequences</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Check if data is a list</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;DNA data must be a list, got </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">data</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Check if data is empty</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;DNA data cannot be empty&quot;</span><span class="p">)</span>

        <span class="c1"># Check if all elements are valid DNA sequences</span>
        <span class="n">valid_nucleotides</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">}</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">data</span><span class="p">):</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="p">(</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="nb">list</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;DNA sequence at index </span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2"> must be a string, &quot;</span>
                               <span class="sa">f</span><span class="s2">&quot;got </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;DNA sequence at index </span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2"> cannot be empty&quot;</span><span class="p">)</span>

            <span class="c1"># Check if all characters are valid nucleotides</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="nb">list</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
                <span class="n">invalid_chars</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
                <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">sequence</span><span class="p">:</span>
                    <span class="n">invalid_chars</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">char</span> <span class="k">for</span> <span class="n">char</span> <span class="ow">in</span> <span class="n">seq</span> <span class="k">if</span> <span class="n">char</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">valid_nucleotides</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">invalid_chars</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">char</span> <span class="k">for</span> <span class="n">char</span> <span class="ow">in</span> <span class="n">sequence</span> <span class="k">if</span> <span class="n">char</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">valid_nucleotides</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">invalid_chars</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;DNA sequence at index </span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2"> contains invalid nucleotides: </span><span class="si">{</span><span class="n">invalid_chars</span><span class="si">}</span><span class="s2">. &quot;</span>
                               <span class="sa">f</span><span class="s2">&quot;Only </span><span class="si">{</span><span class="nb">sorted</span><span class="p">(</span><span class="n">valid_nucleotides</span><span class="p">)</span><span class="si">}</span><span class="s2"> are allowed.&quot;</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_flatten_encoded_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">encoded_data</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Converts nested encoded data structure to flat list of DNA sequences.</span>

<span class="sd">        :param encoded_data: Nested list structure from EncodedData</span>
<span class="sd">        :return: List of DNA sequence strings</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">sequences</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">codeword</span> <span class="ow">in</span> <span class="n">encoded_data</span><span class="p">:</span>
            <span class="c1"># Flatten each codeword and join nucleotides into sequences</span>
            <span class="n">flattened</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_flatten_recursive</span><span class="p">(</span><span class="n">codeword</span><span class="p">)</span>
            <span class="n">sequence</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">flattened</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">sequence</span><span class="p">:</span>  <span class="c1"># Only add non-empty sequences</span>
                <span class="n">sequences</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">sequences</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_flatten_recursive</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">structure</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Recursively flattens nested structure to get nucleotides.</span>

<span class="sd">        :param structure: Nested structure to flatten</span>
<span class="sd">        :return: List of nucleotides</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">result</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">structure</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">structure</span><span class="p">:</span>
                <span class="n">result</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_flatten_recursive</span><span class="p">(</span><span class="n">item</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">structure</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">result</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">get_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get a specific DNA sequence by index.</span>

<span class="sd">        :param index: Index of the sequence</span>
<span class="sd">        :return: The DNA sequence at the specified index</span>
<span class="sd">        :raises IndexError: If index is out of range</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="mi">0</span> <span class="o">&lt;=</span> <span class="n">index</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">IndexError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;DNA sequence index </span><span class="si">{</span><span class="n">index</span><span class="si">}</span><span class="s2"> out of range (0-</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="n">index</span><span class="p">]</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">add_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sequence</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Add a new DNA sequence to the collection.</span>

<span class="sd">        :param sequence: DNA sequence string to add</span>
<span class="sd">        :raises ValueError: If sequence is invalid</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Validate the sequence</span>
        <span class="n">valid_nucleotides</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">}</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;DNA sequence must be a string, got </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;DNA sequence cannot be empty&quot;</span><span class="p">)</span>

        <span class="n">invalid_chars</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">char</span> <span class="k">for</span> <span class="n">char</span> <span class="ow">in</span> <span class="n">sequence</span> <span class="k">if</span> <span class="n">char</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">valid_nucleotides</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">invalid_chars</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;DNA sequence contains invalid nucleotides: </span><span class="si">{</span><span class="n">invalid_chars</span><span class="si">}</span><span class="s2">. &quot;</span>
                           <span class="sa">f</span><span class="s2">&quot;Only </span><span class="si">{</span><span class="nb">sorted</span><span class="p">(</span><span class="n">valid_nucleotides</span><span class="p">)</span><span class="si">}</span><span class="s2"> are allowed.&quot;</span><span class="p">)</span>

        <span class="c1"># Add the sequence and update metrics</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">num_sequences</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">total_length</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">average_length</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">total_length</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_sequences</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">remove_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Remove a DNA sequence by index.</span>

<span class="sd">        :param index: Index of the sequence to remove</span>
<span class="sd">        :raises IndexError: If index is out of range</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="mi">0</span> <span class="o">&lt;=</span> <span class="n">index</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">IndexError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;DNA sequence index </span><span class="si">{</span><span class="n">index</span><span class="si">}</span><span class="s2"> out of range (0-</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">index</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">num_sequences</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">total_length</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="k">else</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">average_length</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">total_length</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_sequences</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_sequences</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">get_sequence_lengths</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get lengths of all DNA sequences.</span>

<span class="sd">        :return: List of sequence lengths</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">]</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">get_nucleotide_counts</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get counts of each nucleotide across all sequences.</span>

<span class="sd">        :return: Dictionary with nucleotide counts</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">counts</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;A&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">}</span>
        <span class="k">for</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">nucleotide</span> <span class="ow">in</span> <span class="n">sequence</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">nucleotide</span> <span class="ow">in</span> <span class="n">counts</span><span class="p">:</span>
                    <span class="n">counts</span><span class="p">[</span><span class="n">nucleotide</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">return</span> <span class="n">counts</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">validate</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Re-validates the current DNA data structure.</span>

<span class="sd">        :return: True if valid</span>
<span class="sd">        :raises: Various exceptions if invalid</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_validate_dna_data</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">True</span>


    <span class="k">def</span><span class="w"> </span><span class="nf">random</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">n</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generate m random nucleotide sequences of length n.</span>

<span class="sd">        Parameters:</span>
<span class="sd">        -----------</span>
<span class="sd">        m : int</span>
<span class="sd">            Number of sequences to generate</span>
<span class="sd">        n : int</span>
<span class="sd">            Length of each sequence</span>

<span class="sd">        Returns:</span>
<span class="sd">        --------</span>
<span class="sd">        InSilicoDNA</span>
<span class="sd">            An InSilicoDNA object containing m random DNA sequences, each of length n</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">m</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Number of sequences (m) must be positive&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">n</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Sequence length (n) must be positive&quot;</span><span class="p">)</span>

        <span class="n">sequences</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">m</span><span class="p">):</span>
            <span class="n">sequence</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="s1">&#39;ACGT&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">))</span>
            <span class="n">sequences</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>

        <span class="n">data</span> <span class="o">=</span> <span class="n">InSilicoDNA</span><span class="p">(</span><span class="n">sequences</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">data</span>


    <span class="k">def</span><span class="w"> </span><span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">output</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;Type: </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">output</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;Number of sequences: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">num_sequences</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">output</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;Total length: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">total_length</span><span class="si">}</span><span class="s2"> nucleotides</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">output</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;Average length: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">average_length</span><span class="si">:</span><span class="s2">.1f</span><span class="si">}</span><span class="s2"> nucleotides</span><span class="se">\n</span><span class="s2">&quot;</span>

        <span class="c1"># Add file paths if they exist</span>
        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;file_paths&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">file_paths</span><span class="p">:</span>
            <span class="n">output</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;File paths: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">file_paths</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>

        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;size&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">size</span><span class="p">:</span>
            <span class="n">output</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;Original size: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">size</span><span class="si">}</span><span class="s2"> bytes</span><span class="se">\n</span><span class="s2">&quot;</span>

        <span class="c1"># Show nucleotide distribution</span>
        <span class="n">counts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_nucleotide_counts</span><span class="p">()</span>
        <span class="n">total_nucleotides</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">counts</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
        <span class="k">if</span> <span class="n">total_nucleotides</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">output</span> <span class="o">+=</span> <span class="s2">&quot;Nucleotide distribution: &quot;</span>
            <span class="n">output</span> <span class="o">+=</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">nt</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">count</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="mi">100</span><span class="o">*</span><span class="n">count</span><span class="o">/</span><span class="n">total_nucleotides</span><span class="si">:</span><span class="s2">.1f</span><span class="si">}</span><span class="s2">%)&quot;</span> 
                              <span class="k">for</span> <span class="n">nt</span><span class="p">,</span> <span class="n">count</span> <span class="ow">in</span> <span class="n">counts</span><span class="o">.</span><span class="n">items</span><span class="p">())</span>
            <span class="n">output</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>

        <span class="n">output</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;DATA: </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)[:</span><span class="mi">100</span><span class="p">]</span><span class="si">}</span><span class="s2">...</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="k">return</span> <span class="n">output</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return the number of DNA sequences.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Allow indexing to get DNA sequences.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">(</span><span class="n">index</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Allow iteration over DNA sequences.&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">iter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



<div class="doc doc-children">










<div class="doc doc-object doc-function">


<h2 id="dnabyte.data_classes.insilicodna.InSilicoDNA.__getitem__" class="doc doc-heading">
            <code class=" language-python"><span class="fm">__getitem__</span><span class="p">(</span><span class="n">index</span><span class="p">)</span></code>

<a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.__getitem__" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Allow indexing to get DNA sequences.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/data_classes/insilicodna.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Allow indexing to get DNA sequences.&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">(</span><span class="n">index</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="dnabyte.data_classes.insilicodna.InSilicoDNA.__init__" class="doc doc-heading">
            <code class=" language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></code>

<a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.__init__" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Creates a DNA object from DNA sequences.</p>
<p>:param data: List of DNA sequences (strings).</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/data_classes/insilicodna.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates a DNA object from DNA sequences.</span>

<span class="sd">    :param data: List of DNA sequences (strings).</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Initialize directly from DNA sequences</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_validate_dna_data</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">data</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">file_paths</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">size</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="c1"># Calculate metrics</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">num_sequences</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="k">else</span> <span class="mi">0</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">total_length</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="k">else</span> <span class="mi">0</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">average_length</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">total_length</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_sequences</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_sequences</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="dnabyte.data_classes.insilicodna.InSilicoDNA.__iter__" class="doc doc-heading">
            <code class=" language-python"><span class="fm">__iter__</span><span class="p">()</span></code>

<a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.__iter__" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Allow iteration over DNA sequences.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/data_classes/insilicodna.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="fm">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Allow iteration over DNA sequences.&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">iter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="dnabyte.data_classes.insilicodna.InSilicoDNA.__len__" class="doc doc-heading">
            <code class=" language-python"><span class="fm">__len__</span><span class="p">()</span></code>

<a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.__len__" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Return the number of DNA sequences.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/data_classes/insilicodna.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Return the number of DNA sequences.&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="dnabyte.data_classes.insilicodna.InSilicoDNA.add_sequence" class="doc doc-heading">
            <code class=" language-python"><span class="n">add_sequence</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span></code>

<a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.add_sequence" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Add a new DNA sequence to the collection.</p>
<p>:param sequence: DNA sequence string to add
:raises ValueError: If sequence is invalid</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/data_classes/insilicodna.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">add_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sequence</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Add a new DNA sequence to the collection.</span>

<span class="sd">    :param sequence: DNA sequence string to add</span>
<span class="sd">    :raises ValueError: If sequence is invalid</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Validate the sequence</span>
    <span class="n">valid_nucleotides</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">}</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;DNA sequence must be a string, got </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;DNA sequence cannot be empty&quot;</span><span class="p">)</span>

    <span class="n">invalid_chars</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">char</span> <span class="k">for</span> <span class="n">char</span> <span class="ow">in</span> <span class="n">sequence</span> <span class="k">if</span> <span class="n">char</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">valid_nucleotides</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">invalid_chars</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;DNA sequence contains invalid nucleotides: </span><span class="si">{</span><span class="n">invalid_chars</span><span class="si">}</span><span class="s2">. &quot;</span>
                       <span class="sa">f</span><span class="s2">&quot;Only </span><span class="si">{</span><span class="nb">sorted</span><span class="p">(</span><span class="n">valid_nucleotides</span><span class="p">)</span><span class="si">}</span><span class="s2"> are allowed.&quot;</span><span class="p">)</span>

    <span class="c1"># Add the sequence and update metrics</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">num_sequences</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">total_length</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">average_length</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">total_length</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_sequences</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="dnabyte.data_classes.insilicodna.InSilicoDNA.get_nucleotide_counts" class="doc doc-heading">
            <code class=" language-python"><span class="n">get_nucleotide_counts</span><span class="p">()</span></code>

<a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.get_nucleotide_counts" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Get counts of each nucleotide across all sequences.</p>
<p>:return: Dictionary with nucleotide counts</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/data_classes/insilicodna.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_nucleotide_counts</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get counts of each nucleotide across all sequences.</span>

<span class="sd">    :return: Dictionary with nucleotide counts</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">counts</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;A&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">}</span>
    <span class="k">for</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">nucleotide</span> <span class="ow">in</span> <span class="n">sequence</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">nucleotide</span> <span class="ow">in</span> <span class="n">counts</span><span class="p">:</span>
                <span class="n">counts</span><span class="p">[</span><span class="n">nucleotide</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="k">return</span> <span class="n">counts</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="dnabyte.data_classes.insilicodna.InSilicoDNA.get_sequence" class="doc doc-heading">
            <code class=" language-python"><span class="n">get_sequence</span><span class="p">(</span><span class="n">index</span><span class="p">)</span></code>

<a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.get_sequence" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Get a specific DNA sequence by index.</p>
<p>:param index: Index of the sequence
:return: The DNA sequence at the specified index
:raises IndexError: If index is out of range</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/data_classes/insilicodna.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get a specific DNA sequence by index.</span>

<span class="sd">    :param index: Index of the sequence</span>
<span class="sd">    :return: The DNA sequence at the specified index</span>
<span class="sd">    :raises IndexError: If index is out of range</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="mi">0</span> <span class="o">&lt;=</span> <span class="n">index</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">IndexError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;DNA sequence index </span><span class="si">{</span><span class="n">index</span><span class="si">}</span><span class="s2"> out of range (0-</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="n">index</span><span class="p">]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="dnabyte.data_classes.insilicodna.InSilicoDNA.get_sequence_lengths" class="doc doc-heading">
            <code class=" language-python"><span class="n">get_sequence_lengths</span><span class="p">()</span></code>

<a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.get_sequence_lengths" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Get lengths of all DNA sequences.</p>
<p>:return: List of sequence lengths</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/data_classes/insilicodna.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_sequence_lengths</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get lengths of all DNA sequences.</span>

<span class="sd">    :return: List of sequence lengths</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="dnabyte.data_classes.insilicodna.InSilicoDNA.random" class="doc doc-heading">
            <code class=" language-python"><span class="n">random</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">n</span><span class="p">)</span></code>

<a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.random" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Generate m random nucleotide sequences of length n.</p>
<h4 id="dnabyte.data_classes.insilicodna.InSilicoDNA.random--parameters">Parameters:<a class="headerlink" href="#dnabyte.data_classes.insilicodna.InSilicoDNA.random--parameters" title="Permanent link">&para;</a></h4>
<p>m : int
    Number of sequences to generate
n : int
    Length of each sequence</p>
<h4 id="dnabyte.data_classes.insilicodna.InSilicoDNA.random--returns">Returns:<a class="headerlink" href="#dnabyte.data_classes.insilicodna.InSilicoDNA.random--returns" title="Permanent link">&para;</a></h4>
<p>InSilicoDNA
    An InSilicoDNA object containing m random DNA sequences, each of length n</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/data_classes/insilicodna.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">random</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">n</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate m random nucleotide sequences of length n.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    -----------</span>
<span class="sd">    m : int</span>
<span class="sd">        Number of sequences to generate</span>
<span class="sd">    n : int</span>
<span class="sd">        Length of each sequence</span>

<span class="sd">    Returns:</span>
<span class="sd">    --------</span>
<span class="sd">    InSilicoDNA</span>
<span class="sd">        An InSilicoDNA object containing m random DNA sequences, each of length n</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">m</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Number of sequences (m) must be positive&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">n</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Sequence length (n) must be positive&quot;</span><span class="p">)</span>

    <span class="n">sequences</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">m</span><span class="p">):</span>
        <span class="n">sequence</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="s1">&#39;ACGT&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">))</span>
        <span class="n">sequences</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>

    <span class="n">data</span> <span class="o">=</span> <span class="n">InSilicoDNA</span><span class="p">(</span><span class="n">sequences</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">data</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="dnabyte.data_classes.insilicodna.InSilicoDNA.remove_sequence" class="doc doc-heading">
            <code class=" language-python"><span class="n">remove_sequence</span><span class="p">(</span><span class="n">index</span><span class="p">)</span></code>

<a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.remove_sequence" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Remove a DNA sequence by index.</p>
<p>:param index: Index of the sequence to remove
:raises IndexError: If index is out of range</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/data_classes/insilicodna.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">remove_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Remove a DNA sequence by index.</span>

<span class="sd">    :param index: Index of the sequence to remove</span>
<span class="sd">    :raises IndexError: If index is out of range</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="mi">0</span> <span class="o">&lt;=</span> <span class="n">index</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">IndexError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;DNA sequence index </span><span class="si">{</span><span class="n">index</span><span class="si">}</span><span class="s2"> out of range (0-</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">index</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">num_sequences</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">total_length</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="k">else</span> <span class="mi">0</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">average_length</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">total_length</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_sequences</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_sequences</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="dnabyte.data_classes.insilicodna.InSilicoDNA.validate" class="doc doc-heading">
            <code class=" language-python"><span class="n">validate</span><span class="p">()</span></code>

<a href="#dnabyte.data_classes.insilicodna.InSilicoDNA.validate" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Re-validates the current DNA data structure.</p>
<p>:return: True if valid
:raises: Various exceptions if invalid</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/data_classes/insilicodna.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">validate</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Re-validates the current DNA data structure.</span>

<span class="sd">    :return: True if valid</span>
<span class="sd">    :raises: Various exceptions if invalid</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_validate_dna_data</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
    <span class="k">return</span> <span class="kc">True</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div><h2 id="core-modules">Core Modules<a class="headerlink" href="#core-modules" title="Permanent link">&para;</a></h2>
<h3 id="params">Params<a class="headerlink" href="#params" title="Permanent link">&para;</a></h3>


<div class="doc doc-object doc-class">



<a id="dnabyte.params.Params"></a>
    <div class="doc doc-contents first">










              <details class="mkdocstrings-source">
                <summary>Source code in <code>dnabyte/params.py</code></summary>
                <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">  4</span>
<span class="normal">  5</span>
<span class="normal">  6</span>
<span class="normal">  7</span>
<span class="normal">  8</span>
<span class="normal">  9</span>
<span class="normal"> 10</span>
<span class="normal"> 11</span>
<span class="normal"> 12</span>
<span class="normal"> 13</span>
<span class="normal"> 14</span>
<span class="normal"> 15</span>
<span class="normal"> 16</span>
<span class="normal"> 17</span>
<span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">Params</span><span class="p">:</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">debug</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>

        <span class="c1"># Set default values for required attributes</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">encoding_method</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">synthesis_method</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">storage_conditions</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sequencing_method</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">binarization_method</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">error_methods</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="c1"># Set parameters from kwargs</span>
        <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="nb">setattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">debug</span> <span class="o">=</span> <span class="n">debug</span>

        <span class="c1"># Load plugins</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">binarization_plugins</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">encoding_plugins</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">storage_plugins</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequencing_plugins</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">error_plugins</span>  <span class="o">=</span> <span class="n">load_plugins</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">binarization_method</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">encoding_method</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">storage_conditions</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequencing_method</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">error_methods</span><span class="p">)</span>

        <span class="c1"># Check binarization parameters</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">binarization_method</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">binarization_method</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">binarization_plugins</span><span class="p">:</span>
            <span class="n">binarization</span> <span class="o">=</span> <span class="n">importlib</span><span class="o">.</span><span class="n">import_module</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;dnabyte.binarization.</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">binarization_method</span><span class="si">}</span><span class="s2">.binarize&quot;</span><span class="p">)</span>
            <span class="n">attributes_bin</span> <span class="o">=</span> <span class="n">binarization</span><span class="o">.</span><span class="n">attributes</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">keys</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">attributes_bin</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="nb">setattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">keys</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid binarization method: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">binarization_method</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="c1"># Check encoding parameters</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">encoding_method</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">encoding_method</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">encoding_plugins</span><span class="p">:</span>
            <span class="n">encoding</span> <span class="o">=</span> <span class="n">importlib</span><span class="o">.</span><span class="n">import_module</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;dnabyte.encoding.</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">encoding_method</span><span class="si">}</span><span class="s2">.encode&quot;</span><span class="p">)</span>
            <span class="n">attributes_enc</span> <span class="o">=</span> <span class="n">encoding</span><span class="o">.</span><span class="n">attributes</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> 
            <span class="k">for</span> <span class="n">keys</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">attributes_enc</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="nb">setattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">keys</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid encoding method: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">encoding_method</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">synthesis_plugins</span> <span class="o">=</span> <span class="n">load_synthesis_plugins</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">synthesis_method</span><span class="p">)</span>

        <span class="c1"># Check synthesis parameters</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">synthesis_method</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">synthesis_method</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">synthesis_plugins</span><span class="p">:</span>
            <span class="n">synthesis</span> <span class="o">=</span> <span class="n">importlib</span><span class="o">.</span><span class="n">import_module</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;dnabyte.synthesis.</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">synthesis_method</span><span class="si">}</span><span class="s2">.synthesize&quot;</span><span class="p">)</span>
            <span class="n">attributes_synth</span> <span class="o">=</span> <span class="n">synthesis</span><span class="o">.</span><span class="n">attributes</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">keys</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">attributes_synth</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="nb">setattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">keys</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid synthesis method: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">synthesis_method</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Check storage parameters</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">storage_conditions</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">storage_conditions</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="n">storage_params_list</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">years_list</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">years</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">years</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">years_list</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">years</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">storage_conditions</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">condition</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">storage_conditions</span><span class="p">):</span>
                <span class="n">dict_of_attributes</span> <span class="o">=</span> <span class="p">{}</span>
                <span class="k">if</span> <span class="n">condition</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">storage_plugins</span><span class="p">:</span>
                    <span class="n">storage</span> <span class="o">=</span> <span class="n">importlib</span><span class="o">.</span><span class="n">import_module</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;dnabyte.storage.</span><span class="si">{</span><span class="n">condition</span><span class="si">}</span><span class="s2">.store&quot;</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">years_list</span><span class="p">):</span>
                        <span class="nb">setattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;years&#39;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">years_list</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="nb">setattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;years&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
                    <span class="n">attributes_store</span> <span class="o">=</span> <span class="n">storage</span><span class="o">.</span><span class="n">attributes</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
                    <span class="nb">delattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;years&#39;</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">keys</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">attributes_store</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                        <span class="n">dict_of_attributes</span><span class="p">[</span><span class="n">keys</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>
                    <span class="n">storage_params_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dict_of_attributes</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid storage condition: </span><span class="si">{</span><span class="n">condition</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">storage_params_list</span> <span class="o">=</span> <span class="n">storage_params_list</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">storage_conditions</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">storage_conditions</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">storage_plugins</span><span class="p">:</span>

            <span class="n">dict_of_attributes</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">storage</span> <span class="o">=</span> <span class="n">importlib</span><span class="o">.</span><span class="n">import_module</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;dnabyte.storage.</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">storage_conditions</span><span class="si">}</span><span class="s2">.store&quot;</span><span class="p">)</span>
            <span class="n">attributes_store</span> <span class="o">=</span> <span class="n">storage</span><span class="o">.</span><span class="n">attributes</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">keys</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">attributes_store</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="n">dict_of_attributes</span><span class="p">[</span><span class="n">keys</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">storage_params</span> <span class="o">=</span> <span class="n">dict_of_attributes</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid storage conditions: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">storage_conditions</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Check error parameters</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">error_methods</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">error_methods</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="n">error_params_list</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">condition</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">error_methods</span><span class="p">):</span>
                <span class="n">dict_of_attributes</span> <span class="o">=</span> <span class="p">{}</span>
                <span class="k">if</span> <span class="n">condition</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">error_plugins</span><span class="p">:</span>
                    <span class="n">errs</span> <span class="o">=</span> <span class="n">importlib</span><span class="o">.</span><span class="n">import_module</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;dnabyte.misc_errors.</span><span class="si">{</span><span class="n">condition</span><span class="si">}</span><span class="s2">.err&quot;</span><span class="p">)</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;error_params_temp&#39;</span><span class="p">):</span>
                        <span class="k">pass</span>
                    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">error_params_temp</span><span class="p">,</span> <span class="nb">dict</span><span class="p">):</span>
                        <span class="k">if</span> <span class="n">condition</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">error_params_temp</span><span class="p">:</span>
                            <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">error_params_temp</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                                <span class="nb">setattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span>
                        <span class="nb">delattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;error_params_temp&#39;</span><span class="p">)</span>
                    <span class="n">attributes_store</span> <span class="o">=</span> <span class="n">errs</span><span class="o">.</span><span class="n">attributes</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">keys</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">attributes_store</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                        <span class="n">dict_of_attributes</span><span class="p">[</span><span class="n">keys</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>
                    <span class="n">error_params_list</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="n">condition</span><span class="p">:</span> <span class="n">dict_of_attributes</span><span class="p">})</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid storage condition: </span><span class="si">{</span><span class="n">condition</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="nb">setattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;error_params_list&#39;</span><span class="p">,</span> <span class="n">error_params_list</span><span class="p">)</span>
            <span class="nb">delattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;error_params&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">error_methods</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">error_methods</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">error_plugins</span><span class="p">:</span>

            <span class="n">dict_of_attributes</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">errs</span> <span class="o">=</span> <span class="n">importlib</span><span class="o">.</span><span class="n">import_module</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;dnabyte.misc_errors.</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">error_methods</span><span class="si">}</span><span class="s2">.err&quot;</span><span class="p">)</span>
            <span class="n">attributes_err</span> <span class="o">=</span> <span class="n">errs</span><span class="o">.</span><span class="n">attributes</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">keys</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">attributes_err</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="n">dict_of_attributes</span><span class="p">[</span><span class="n">keys</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>
            <span class="nb">setattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;error_params&#39;</span><span class="p">,</span> <span class="n">dict_of_attributes</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid error methods: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">error_methods</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="c1"># Check sequencing parameters</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequencing_method</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequencing_method</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequencing_plugins</span><span class="p">:</span>
            <span class="n">sequencing</span> <span class="o">=</span> <span class="n">importlib</span><span class="o">.</span><span class="n">import_module</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;dnabyte.sequencing.</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">sequencing_method</span><span class="si">}</span><span class="s2">.sequence&quot;</span><span class="p">)</span>
            <span class="n">attributes_seq</span> <span class="o">=</span> <span class="n">sequencing</span><span class="o">.</span><span class="n">attributes</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">keys</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">attributes_seq</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="nb">setattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">keys</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid sequencing method: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">sequencing_method</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>


    <span class="k">def</span><span class="w"> </span><span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;Params:</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__dict__</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">output</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;  </span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">value</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="k">return</span> <span class="n">output</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">params_range</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Alternative constructor that creates a list of Params instances for each item in the list if one of the </span>
<span class="sd">        parameters is a non-empty list.</span>

<span class="sd">        Parameters:</span>
<span class="sd">        kwargs (dict): The parameters for the Params instances.</span>

<span class="sd">        Returns:</span>
<span class="sd">        list: A list of Params instances.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">params_list</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># Attributes that should not be treated as range parameters (must remain as lists)</span>
        <span class="n">excluded_attrs</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;file_paths&#39;</span><span class="p">}</span>

        <span class="c1"># Find the first attribute that is a non-empty list (excluding special attributes)</span>
        <span class="k">for</span> <span class="n">attr_name</span><span class="p">,</span> <span class="n">attr_value</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">attr_name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">excluded_attrs</span> <span class="ow">and</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">attr_value</span><span class="p">,</span> <span class="nb">list</span><span class="p">)</span> <span class="ow">and</span> <span class="n">attr_value</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">attr_value</span><span class="p">:</span>
                    <span class="n">new_kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="o">**</span><span class="n">kwargs</span><span class="p">,</span> <span class="n">attr_name</span><span class="p">:</span> <span class="n">item</span><span class="p">}</span>
                    <span class="n">params_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">cls</span><span class="p">(</span><span class="o">**</span><span class="n">new_kwargs</span><span class="p">))</span>
                <span class="k">break</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># If no attribute is a non-empty list, return a list with a single instance</span>
            <span class="n">params_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">cls</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">))</span>

        <span class="k">return</span> <span class="n">params_list</span>
</code></pre></div></td></tr></table></div>
              </details>



<div class="doc doc-children">










<div class="doc doc-object doc-function">


<h2 id="dnabyte.params.Params.params_range" class="doc doc-heading">
            <code class=" language-python"><span class="n">params_range</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

<a href="#dnabyte.params.Params.params_range" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Alternative constructor that creates a list of Params instances for each item in the list if one of the 
parameters is a non-empty list.</p>
<p>Parameters:
kwargs (dict): The parameters for the Params instances.</p>
<p>Returns:
list: A list of Params instances.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/params.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">params_range</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Alternative constructor that creates a list of Params instances for each item in the list if one of the </span>
<span class="sd">    parameters is a non-empty list.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    kwargs (dict): The parameters for the Params instances.</span>

<span class="sd">    Returns:</span>
<span class="sd">    list: A list of Params instances.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">params_list</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># Attributes that should not be treated as range parameters (must remain as lists)</span>
    <span class="n">excluded_attrs</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;file_paths&#39;</span><span class="p">}</span>

    <span class="c1"># Find the first attribute that is a non-empty list (excluding special attributes)</span>
    <span class="k">for</span> <span class="n">attr_name</span><span class="p">,</span> <span class="n">attr_value</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">attr_name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">excluded_attrs</span> <span class="ow">and</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">attr_value</span><span class="p">,</span> <span class="nb">list</span><span class="p">)</span> <span class="ow">and</span> <span class="n">attr_value</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">attr_value</span><span class="p">:</span>
                <span class="n">new_kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="o">**</span><span class="n">kwargs</span><span class="p">,</span> <span class="n">attr_name</span><span class="p">:</span> <span class="n">item</span><span class="p">}</span>
                <span class="n">params_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">cls</span><span class="p">(</span><span class="o">**</span><span class="n">new_kwargs</span><span class="p">))</span>
            <span class="k">break</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># If no attribute is a non-empty list, return a list with a single instance</span>
        <span class="n">params_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">cls</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">params_list</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div><h3 id="library">Library<a class="headerlink" href="#library" title="Permanent link">&para;</a></h3>


<div class="doc doc-object doc-class">



<a id="dnabyte.library.Library"></a>
    <div class="doc doc-contents first">



        <p>A class to represent a DNA oligo library.</p>








              <details class="mkdocstrings-source">
                <summary>Source code in <code>dnabyte/library.py</code></summary>
                <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">  3</span>
<span class="normal">  4</span>
<span class="normal">  5</span>
<span class="normal">  6</span>
<span class="normal">  7</span>
<span class="normal">  8</span>
<span class="normal">  9</span>
<span class="normal"> 10</span>
<span class="normal"> 11</span>
<span class="normal"> 12</span>
<span class="normal"> 13</span>
<span class="normal"> 14</span>
<span class="normal"> 15</span>
<span class="normal"> 16</span>
<span class="normal"> 17</span>
<span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">Library</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A class to represent a DNA oligo library.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">structure</span><span class="o">=</span><span class="s1">&#39;linear_assembly&#39;</span><span class="p">,</span> <span class="n">library</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialize a Library object.</span>

<span class="sd">        :param structure: The structure of the library indicated by a string &#39;simple&#39; (default) or &#39;poly&#39;.</span>
<span class="sd">        :param library: The library data, default is None.</span>
<span class="sd">        :param kwargs: Additional keyword arguments.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">structure</span> <span class="o">=</span> <span class="n">structure</span>

        <span class="k">if</span> <span class="n">library</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">library</span> <span class="o">=</span> <span class="n">library</span>
        <span class="k">elif</span> <span class="s1">&#39;filename&#39;</span> <span class="ow">in</span> <span class="n">kwargs</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span> <span class="o">==</span> <span class="s1">&#39;linear_assembly&#39;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">library</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">read_library</span><span class="p">(</span><span class="n">kwargs</span><span class="p">[</span><span class="s1">&#39;filename&#39;</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">leftmotives</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">rightmotives</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">dictmotives</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">translationlibleft</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">translationlibright</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">motive_pairs</span><span class="p">()</span>


        <span class="k">elif</span> <span class="s1">&#39;filename&#39;</span> <span class="ow">in</span> <span class="n">kwargs</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span> <span class="o">==</span> <span class="s1">&#39;positional_assembly&#39;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">messages</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">generic</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">position</span>  <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">read_library_poly</span><span class="p">(</span><span class="n">kwargs</span><span class="p">[</span><span class="s1">&#39;filename&#39;</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">library</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">messages</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">generic</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">position</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">messageleft</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">messageright</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">genericlib</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">connectorlib</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">dictmotives</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">motive_pairs_poly</span><span class="p">()</span>


    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">read_library</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">file_path</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Read the library data from a file.</span>

<span class="sd">        :param file_path: Path to the file containing the library data.</span>
<span class="sd">        :return: A list of DNA sequences.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># TODO: Fallunterscheidung mit fasta und .csv</span>
        <span class="c1"># TODO: make it an actual class method</span>
        <span class="c1"># TODO: check if read in library is correct</span>


        <span class="n">DNAs</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">csv_file</span><span class="p">:</span>
            <span class="n">csv_reader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">reader</span><span class="p">(</span><span class="n">csv_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">csv_reader</span><span class="p">:</span>    
                <span class="n">DNAs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">DNAs</span>      

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">read_library_poly</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">file_path</span><span class="p">):</span>

        <span class="c1"># TODO: check if read in library is correct</span>

        <span class="n">DNAs</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">csv_file</span><span class="p">:</span>
            <span class="n">csv_reader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">reader</span><span class="p">(</span><span class="n">csv_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">csv_reader</span><span class="p">:</span>    
                <span class="n">DNAs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

        <span class="n">messages</span> <span class="o">=</span> <span class="n">DNAs</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="n">DNAs</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="s1">&#39;Generic&#39;</span><span class="p">)]</span>
        <span class="n">generic</span> <span class="o">=</span> <span class="n">DNAs</span><span class="p">[</span><span class="n">DNAs</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="s1">&#39;Generic&#39;</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">:</span><span class="n">DNAs</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="s1">&#39;Connector&#39;</span><span class="p">)]</span>
        <span class="n">positions</span> <span class="o">=</span> <span class="n">DNAs</span><span class="p">[</span><span class="n">DNAs</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="s1">&#39;Connector&#39;</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">:]</span>

        <span class="k">return</span> <span class="n">messages</span><span class="p">,</span>  <span class="n">generic</span><span class="p">,</span> <span class="n">positions</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">complementmap</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">string</span><span class="p">):</span>
        <span class="n">compliment</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">string</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="s1">&#39;A&#39;</span><span class="p">:</span>
                <span class="n">compliment</span> <span class="o">+=</span> <span class="s1">&#39;T&#39;</span>
            <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="s1">&#39;T&#39;</span><span class="p">:</span>
                <span class="n">compliment</span> <span class="o">+=</span> <span class="s1">&#39;A&#39;</span>
            <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="s1">&#39;C&#39;</span><span class="p">:</span>
                <span class="n">compliment</span> <span class="o">+=</span> <span class="s1">&#39;G&#39;</span>
            <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="s1">&#39;G&#39;</span><span class="p">:</span>
                <span class="n">compliment</span> <span class="o">+=</span> <span class="s1">&#39;C&#39;</span>
        <span class="n">compliment</span> <span class="o">=</span> <span class="n">compliment</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">compliment</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">motive_pairs</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generate motive pairs from the library data.</span>

<span class="sd">        :param library: The library data.</span>
<span class="sd">        :return: A tuple containing left motifs, right motifs, a dictionary of motives,</span>
<span class="sd">             translation library for left motifs, and translation library for right motifs.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">DNAs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">library</span>
        <span class="n">oligolen</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">DNAs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="n">oligolenhalf</span> <span class="o">=</span> <span class="n">oligolen</span> <span class="o">//</span> <span class="mi">2</span>
        <span class="n">leftmotifs</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">rightmotifs</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">dna</span> <span class="ow">in</span> <span class="n">DNAs</span><span class="p">:</span>
            <span class="n">left_motif</span> <span class="o">=</span> <span class="n">dna</span><span class="p">[:</span><span class="n">oligolenhalf</span><span class="p">]</span>
            <span class="n">right_motif</span> <span class="o">=</span> <span class="n">dna</span><span class="p">[</span><span class="n">oligolenhalf</span><span class="p">:]</span>

            <span class="k">if</span> <span class="n">left_motif</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">leftmotifs</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">complementmap</span><span class="p">(</span><span class="n">left_motif</span><span class="p">)</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">leftmotifs</span><span class="p">:</span>
                <span class="n">leftmotifs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">left_motif</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">right_motif</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">rightmotifs</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">complementmap</span><span class="p">(</span><span class="n">right_motif</span><span class="p">)</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">rightmotifs</span><span class="p">:</span>
                <span class="n">rightmotifs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">right_motif</span><span class="p">)</span> 

        <span class="n">leftmotifs</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
        <span class="n">rightmotifs</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

        <span class="n">translationlibleft</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">translationlibright</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">dictmotives</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">motif</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">leftmotifs</span><span class="p">):</span>
            <span class="n">translationlibleft</span><span class="p">[</span><span class="n">motif</span><span class="p">]</span> <span class="o">=</span> <span class="nb">chr</span><span class="p">(</span><span class="mi">97</span> <span class="o">+</span> <span class="n">i</span><span class="p">)</span>
            <span class="n">translationlibleft</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">complementmap</span><span class="p">(</span><span class="n">motif</span><span class="p">)]</span> <span class="o">=</span> <span class="nb">chr</span><span class="p">(</span><span class="mi">97</span> <span class="o">+</span> <span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span>
            <span class="n">dictmotives</span><span class="p">[</span><span class="nb">chr</span><span class="p">(</span><span class="mi">97</span> <span class="o">+</span> <span class="n">i</span><span class="p">)]</span> <span class="o">=</span> <span class="nb">chr</span><span class="p">(</span><span class="mi">97</span> <span class="o">+</span> <span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span>
            <span class="n">dictmotives</span><span class="p">[</span><span class="nb">chr</span><span class="p">(</span><span class="mi">97</span> <span class="o">+</span> <span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">chr</span><span class="p">(</span><span class="mi">97</span> <span class="o">+</span> <span class="n">i</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">motif</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">rightmotifs</span><span class="p">):</span>
            <span class="n">translationlibright</span><span class="p">[</span><span class="n">motif</span><span class="p">]</span> <span class="o">=</span> <span class="nb">chr</span><span class="p">(</span><span class="mi">65</span> <span class="o">+</span> <span class="n">i</span><span class="p">)</span>
            <span class="n">translationlibright</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">complementmap</span><span class="p">(</span><span class="n">motif</span><span class="p">)]</span> <span class="o">=</span> <span class="nb">chr</span><span class="p">(</span><span class="mi">65</span> <span class="o">+</span> <span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span>
            <span class="n">dictmotives</span><span class="p">[</span><span class="nb">chr</span><span class="p">(</span><span class="mi">65</span> <span class="o">+</span> <span class="n">i</span><span class="p">)]</span> <span class="o">=</span> <span class="nb">chr</span><span class="p">(</span><span class="mi">65</span> <span class="o">+</span> <span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span>
            <span class="n">dictmotives</span><span class="p">[</span><span class="nb">chr</span><span class="p">(</span><span class="mi">65</span> <span class="o">+</span> <span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">chr</span><span class="p">(</span><span class="mi">65</span> <span class="o">+</span> <span class="n">i</span><span class="p">)</span>

        <span class="n">dictmotives</span><span class="p">[</span><span class="s1">&#39;empty&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;empty*&#39;</span>
        <span class="n">dictmotives</span><span class="p">[</span><span class="s1">&#39;empty*&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;empty&#39;</span>
        <span class="n">dictmotives</span><span class="p">[</span><span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="k">return</span> <span class="n">leftmotifs</span><span class="p">,</span> <span class="n">rightmotifs</span><span class="p">,</span> <span class="n">dictmotives</span><span class="p">,</span> <span class="n">translationlibleft</span><span class="p">,</span> <span class="n">translationlibright</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">motive_pairs_poly</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generate motive pairs from the library data.</span>

<span class="sd">        :param library: The library data.</span>
<span class="sd">        :return: A tuple containing left motifs, right motifs, a dictionary of motives,</span>
<span class="sd">             translation library for left motifs, and translation library for right motifs.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">messagelibleft</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">message_libright</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">genericlib</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">connectorlib</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">dictmotives</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">messagesleft</span> <span class="o">=</span><span class="p">[]</span>
        <span class="n">messagesright</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">messages</span><span class="p">)):</span>
            <span class="n">messagesleft</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">messages</span><span class="p">[</span><span class="n">i</span><span class="p">][:</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">messages</span><span class="p">[</span><span class="n">i</span><span class="p">])</span><span class="o">//</span><span class="mi">2</span><span class="p">])</span>
            <span class="n">messagesright</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">messages</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">messages</span><span class="p">[</span><span class="n">i</span><span class="p">])</span><span class="o">//</span><span class="mi">2</span><span class="p">:])</span>
        <span class="c1"># messagesleft = list(set(messagesleft))</span>
        <span class="c1"># messagesright = list(set(messagesright))</span>
        <span class="c1"># messagesleft.sort()</span>
        <span class="c1"># messagesright.sort()</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">motif</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">messagesleft</span><span class="p">):</span>
            <span class="n">messagelibleft</span><span class="p">[</span><span class="n">motif</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;ml&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
            <span class="n">messagelibleft</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">complementmap</span><span class="p">(</span><span class="n">motif</span><span class="p">)]</span> <span class="o">=</span> <span class="s1">&#39;ml&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span>
            <span class="n">dictmotives</span><span class="p">[</span><span class="s1">&#39;ml&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)]</span> <span class="o">=</span> <span class="s1">&#39;ml&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span>
            <span class="n">dictmotives</span><span class="p">[</span><span class="s1">&#39;ml&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;ml&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">motif</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">messagesright</span><span class="p">):</span>
            <span class="n">message_libright</span><span class="p">[</span><span class="n">motif</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;mr&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
            <span class="n">message_libright</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">complementmap</span><span class="p">(</span><span class="n">motif</span><span class="p">)]</span> <span class="o">=</span> <span class="s1">&#39;mr&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span>
            <span class="n">dictmotives</span><span class="p">[</span><span class="s1">&#39;mr&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)]</span> <span class="o">=</span> <span class="s1">&#39;mr&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span>
            <span class="n">dictmotives</span><span class="p">[</span><span class="s1">&#39;mr&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;mr&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">motif</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">generic</span><span class="p">):</span>
            <span class="n">genericlib</span><span class="p">[</span><span class="n">motif</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;g&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
            <span class="n">genericlib</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">complementmap</span><span class="p">(</span><span class="n">motif</span><span class="p">)]</span> <span class="o">=</span> <span class="s1">&#39;g&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span>
            <span class="n">dictmotives</span><span class="p">[</span><span class="s1">&#39;g&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)]</span> <span class="o">=</span> <span class="s1">&#39;g&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span>
            <span class="n">dictmotives</span><span class="p">[</span><span class="s1">&#39;g&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;g&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">motif</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">position</span><span class="p">):</span>
            <span class="n">connectorlib</span><span class="p">[</span><span class="n">motif</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;c&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
            <span class="n">connectorlib</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">complementmap</span><span class="p">(</span><span class="n">motif</span><span class="p">)]</span> <span class="o">=</span> <span class="s1">&#39;c&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="o">+</span> <span class="s1">&#39;*&#39;</span>
            <span class="n">dictmotives</span><span class="p">[</span><span class="s1">&#39;c&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)]</span> <span class="o">=</span> <span class="s1">&#39;c&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span>
            <span class="n">dictmotives</span><span class="p">[</span><span class="s1">&#39;c&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;c&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

        <span class="n">dictmotives</span><span class="p">[</span><span class="s1">&#39;empty&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;empty*&#39;</span>
        <span class="n">dictmotives</span><span class="p">[</span><span class="s1">&#39;empty*&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;empty&#39;</span>
        <span class="n">dictmotives</span><span class="p">[</span><span class="kc">None</span><span class="p">]</span><span class="o">=</span> <span class="kc">None</span>

        <span class="k">return</span> <span class="n">messagelibleft</span><span class="p">,</span> <span class="n">message_libright</span><span class="p">,</span> <span class="n">genericlib</span><span class="p">,</span> <span class="n">connectorlib</span><span class="p">,</span> <span class="n">dictmotives</span>
</code></pre></div></td></tr></table></div>
              </details>



<div class="doc doc-children">










<div class="doc doc-object doc-function">


<h2 id="dnabyte.library.Library.__init__" class="doc doc-heading">
            <code class=" language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">structure</span><span class="o">=</span><span class="s1">&#39;linear_assembly&#39;</span><span class="p">,</span> <span class="n">library</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

<a href="#dnabyte.library.Library.__init__" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Initialize a Library object.</p>
<p>:param structure: The structure of the library indicated by a string 'simple' (default) or 'poly'.
:param library: The library data, default is None.
:param kwargs: Additional keyword arguments.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/library.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">structure</span><span class="o">=</span><span class="s1">&#39;linear_assembly&#39;</span><span class="p">,</span> <span class="n">library</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Initialize a Library object.</span>

<span class="sd">    :param structure: The structure of the library indicated by a string &#39;simple&#39; (default) or &#39;poly&#39;.</span>
<span class="sd">    :param library: The library data, default is None.</span>
<span class="sd">    :param kwargs: Additional keyword arguments.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">structure</span> <span class="o">=</span> <span class="n">structure</span>

    <span class="k">if</span> <span class="n">library</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">library</span> <span class="o">=</span> <span class="n">library</span>
    <span class="k">elif</span> <span class="s1">&#39;filename&#39;</span> <span class="ow">in</span> <span class="n">kwargs</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span> <span class="o">==</span> <span class="s1">&#39;linear_assembly&#39;</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">library</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">read_library</span><span class="p">(</span><span class="n">kwargs</span><span class="p">[</span><span class="s1">&#39;filename&#39;</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">leftmotives</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">rightmotives</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">dictmotives</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">translationlibleft</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">translationlibright</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">motive_pairs</span><span class="p">()</span>


    <span class="k">elif</span> <span class="s1">&#39;filename&#39;</span> <span class="ow">in</span> <span class="n">kwargs</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">structure</span> <span class="o">==</span> <span class="s1">&#39;positional_assembly&#39;</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">messages</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">generic</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">position</span>  <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">read_library_poly</span><span class="p">(</span><span class="n">kwargs</span><span class="p">[</span><span class="s1">&#39;filename&#39;</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">library</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">messages</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">generic</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">position</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">messageleft</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">messageright</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">genericlib</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">connectorlib</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">dictmotives</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">motive_pairs_poly</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="dnabyte.library.Library.motive_pairs" class="doc doc-heading">
            <code class=" language-python"><span class="n">motive_pairs</span><span class="p">()</span></code>

<a href="#dnabyte.library.Library.motive_pairs" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Generate motive pairs from the library data.</p>
<p>:param library: The library data.
:return: A tuple containing left motifs, right motifs, a dictionary of motives,
     translation library for left motifs, and translation library for right motifs.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/library.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">motive_pairs</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate motive pairs from the library data.</span>

<span class="sd">    :param library: The library data.</span>
<span class="sd">    :return: A tuple containing left motifs, right motifs, a dictionary of motives,</span>
<span class="sd">         translation library for left motifs, and translation library for right motifs.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">DNAs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">library</span>
    <span class="n">oligolen</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">DNAs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">oligolenhalf</span> <span class="o">=</span> <span class="n">oligolen</span> <span class="o">//</span> <span class="mi">2</span>
    <span class="n">leftmotifs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">rightmotifs</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">dna</span> <span class="ow">in</span> <span class="n">DNAs</span><span class="p">:</span>
        <span class="n">left_motif</span> <span class="o">=</span> <span class="n">dna</span><span class="p">[:</span><span class="n">oligolenhalf</span><span class="p">]</span>
        <span class="n">right_motif</span> <span class="o">=</span> <span class="n">dna</span><span class="p">[</span><span class="n">oligolenhalf</span><span class="p">:]</span>

        <span class="k">if</span> <span class="n">left_motif</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">leftmotifs</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">complementmap</span><span class="p">(</span><span class="n">left_motif</span><span class="p">)</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">leftmotifs</span><span class="p">:</span>
            <span class="n">leftmotifs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">left_motif</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">right_motif</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">rightmotifs</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">complementmap</span><span class="p">(</span><span class="n">right_motif</span><span class="p">)</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">rightmotifs</span><span class="p">:</span>
            <span class="n">rightmotifs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">right_motif</span><span class="p">)</span> 

    <span class="n">leftmotifs</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
    <span class="n">rightmotifs</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

    <span class="n">translationlibleft</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">translationlibright</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">dictmotives</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">motif</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">leftmotifs</span><span class="p">):</span>
        <span class="n">translationlibleft</span><span class="p">[</span><span class="n">motif</span><span class="p">]</span> <span class="o">=</span> <span class="nb">chr</span><span class="p">(</span><span class="mi">97</span> <span class="o">+</span> <span class="n">i</span><span class="p">)</span>
        <span class="n">translationlibleft</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">complementmap</span><span class="p">(</span><span class="n">motif</span><span class="p">)]</span> <span class="o">=</span> <span class="nb">chr</span><span class="p">(</span><span class="mi">97</span> <span class="o">+</span> <span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span>
        <span class="n">dictmotives</span><span class="p">[</span><span class="nb">chr</span><span class="p">(</span><span class="mi">97</span> <span class="o">+</span> <span class="n">i</span><span class="p">)]</span> <span class="o">=</span> <span class="nb">chr</span><span class="p">(</span><span class="mi">97</span> <span class="o">+</span> <span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span>
        <span class="n">dictmotives</span><span class="p">[</span><span class="nb">chr</span><span class="p">(</span><span class="mi">97</span> <span class="o">+</span> <span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">chr</span><span class="p">(</span><span class="mi">97</span> <span class="o">+</span> <span class="n">i</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">motif</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">rightmotifs</span><span class="p">):</span>
        <span class="n">translationlibright</span><span class="p">[</span><span class="n">motif</span><span class="p">]</span> <span class="o">=</span> <span class="nb">chr</span><span class="p">(</span><span class="mi">65</span> <span class="o">+</span> <span class="n">i</span><span class="p">)</span>
        <span class="n">translationlibright</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">complementmap</span><span class="p">(</span><span class="n">motif</span><span class="p">)]</span> <span class="o">=</span> <span class="nb">chr</span><span class="p">(</span><span class="mi">65</span> <span class="o">+</span> <span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span>
        <span class="n">dictmotives</span><span class="p">[</span><span class="nb">chr</span><span class="p">(</span><span class="mi">65</span> <span class="o">+</span> <span class="n">i</span><span class="p">)]</span> <span class="o">=</span> <span class="nb">chr</span><span class="p">(</span><span class="mi">65</span> <span class="o">+</span> <span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span>
        <span class="n">dictmotives</span><span class="p">[</span><span class="nb">chr</span><span class="p">(</span><span class="mi">65</span> <span class="o">+</span> <span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">chr</span><span class="p">(</span><span class="mi">65</span> <span class="o">+</span> <span class="n">i</span><span class="p">)</span>

    <span class="n">dictmotives</span><span class="p">[</span><span class="s1">&#39;empty&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;empty*&#39;</span>
    <span class="n">dictmotives</span><span class="p">[</span><span class="s1">&#39;empty*&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;empty&#39;</span>
    <span class="n">dictmotives</span><span class="p">[</span><span class="kc">None</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">return</span> <span class="n">leftmotifs</span><span class="p">,</span> <span class="n">rightmotifs</span><span class="p">,</span> <span class="n">dictmotives</span><span class="p">,</span> <span class="n">translationlibleft</span><span class="p">,</span> <span class="n">translationlibright</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="dnabyte.library.Library.motive_pairs_poly" class="doc doc-heading">
            <code class=" language-python"><span class="n">motive_pairs_poly</span><span class="p">()</span></code>

<a href="#dnabyte.library.Library.motive_pairs_poly" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Generate motive pairs from the library data.</p>
<p>:param library: The library data.
:return: A tuple containing left motifs, right motifs, a dictionary of motives,
     translation library for left motifs, and translation library for right motifs.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/library.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">motive_pairs_poly</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate motive pairs from the library data.</span>

<span class="sd">    :param library: The library data.</span>
<span class="sd">    :return: A tuple containing left motifs, right motifs, a dictionary of motives,</span>
<span class="sd">         translation library for left motifs, and translation library for right motifs.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">messagelibleft</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">message_libright</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">genericlib</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">connectorlib</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">dictmotives</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">messagesleft</span> <span class="o">=</span><span class="p">[]</span>
    <span class="n">messagesright</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">messages</span><span class="p">)):</span>
        <span class="n">messagesleft</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">messages</span><span class="p">[</span><span class="n">i</span><span class="p">][:</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">messages</span><span class="p">[</span><span class="n">i</span><span class="p">])</span><span class="o">//</span><span class="mi">2</span><span class="p">])</span>
        <span class="n">messagesright</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">messages</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">messages</span><span class="p">[</span><span class="n">i</span><span class="p">])</span><span class="o">//</span><span class="mi">2</span><span class="p">:])</span>
    <span class="c1"># messagesleft = list(set(messagesleft))</span>
    <span class="c1"># messagesright = list(set(messagesright))</span>
    <span class="c1"># messagesleft.sort()</span>
    <span class="c1"># messagesright.sort()</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">motif</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">messagesleft</span><span class="p">):</span>
        <span class="n">messagelibleft</span><span class="p">[</span><span class="n">motif</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;ml&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        <span class="n">messagelibleft</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">complementmap</span><span class="p">(</span><span class="n">motif</span><span class="p">)]</span> <span class="o">=</span> <span class="s1">&#39;ml&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span>
        <span class="n">dictmotives</span><span class="p">[</span><span class="s1">&#39;ml&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)]</span> <span class="o">=</span> <span class="s1">&#39;ml&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span>
        <span class="n">dictmotives</span><span class="p">[</span><span class="s1">&#39;ml&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;ml&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">motif</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">messagesright</span><span class="p">):</span>
        <span class="n">message_libright</span><span class="p">[</span><span class="n">motif</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;mr&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        <span class="n">message_libright</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">complementmap</span><span class="p">(</span><span class="n">motif</span><span class="p">)]</span> <span class="o">=</span> <span class="s1">&#39;mr&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span>
        <span class="n">dictmotives</span><span class="p">[</span><span class="s1">&#39;mr&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)]</span> <span class="o">=</span> <span class="s1">&#39;mr&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span>
        <span class="n">dictmotives</span><span class="p">[</span><span class="s1">&#39;mr&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;mr&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">motif</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">generic</span><span class="p">):</span>
        <span class="n">genericlib</span><span class="p">[</span><span class="n">motif</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;g&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        <span class="n">genericlib</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">complementmap</span><span class="p">(</span><span class="n">motif</span><span class="p">)]</span> <span class="o">=</span> <span class="s1">&#39;g&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span>
        <span class="n">dictmotives</span><span class="p">[</span><span class="s1">&#39;g&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)]</span> <span class="o">=</span> <span class="s1">&#39;g&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span>
        <span class="n">dictmotives</span><span class="p">[</span><span class="s1">&#39;g&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;g&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">motif</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">position</span><span class="p">):</span>
        <span class="n">connectorlib</span><span class="p">[</span><span class="n">motif</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;c&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        <span class="n">connectorlib</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">complementmap</span><span class="p">(</span><span class="n">motif</span><span class="p">)]</span> <span class="o">=</span> <span class="s1">&#39;c&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="o">+</span> <span class="s1">&#39;*&#39;</span>
        <span class="n">dictmotives</span><span class="p">[</span><span class="s1">&#39;c&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)]</span> <span class="o">=</span> <span class="s1">&#39;c&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span>
        <span class="n">dictmotives</span><span class="p">[</span><span class="s1">&#39;c&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;*&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;c&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

    <span class="n">dictmotives</span><span class="p">[</span><span class="s1">&#39;empty&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;empty*&#39;</span>
    <span class="n">dictmotives</span><span class="p">[</span><span class="s1">&#39;empty*&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;empty&#39;</span>
    <span class="n">dictmotives</span><span class="p">[</span><span class="kc">None</span><span class="p">]</span><span class="o">=</span> <span class="kc">None</span>

    <span class="k">return</span> <span class="n">messagelibleft</span><span class="p">,</span> <span class="n">message_libright</span><span class="p">,</span> <span class="n">genericlib</span><span class="p">,</span> <span class="n">connectorlib</span><span class="p">,</span> <span class="n">dictmotives</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="dnabyte.library.Library.read_library" class="doc doc-heading">
            <code class=" language-python"><span class="n">read_library</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

<a href="#dnabyte.library.Library.read_library" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Read the library data from a file.</p>
<p>:param file_path: Path to the file containing the library data.
:return: A list of DNA sequences.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/library.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">read_library</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">file_path</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Read the library data from a file.</span>

<span class="sd">    :param file_path: Path to the file containing the library data.</span>
<span class="sd">    :return: A list of DNA sequences.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># TODO: Fallunterscheidung mit fasta und .csv</span>
    <span class="c1"># TODO: make it an actual class method</span>
    <span class="c1"># TODO: check if read in library is correct</span>


    <span class="n">DNAs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">csv_file</span><span class="p">:</span>
        <span class="n">csv_reader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">reader</span><span class="p">(</span><span class="n">csv_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">csv_reader</span><span class="p">:</span>    
            <span class="n">DNAs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">DNAs</span>      
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div><h3 id="oligo">Oligo<a class="headerlink" href="#oligo" title="Permanent link">&para;</a></h3>


<div class="doc doc-object doc-class">



<a id="dnabyte.oligo.Oligo"></a>
    <div class="doc doc-contents first">



        <p>This class models a single or double stranded oligonucleotide. It is always instanciated as single stranded,
from which double stranded oligonucleotides can be created by hybridising it with another oligo.</p>


<p><span class="doc-section-title">Attributes:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code><span title="dnabyte.oligo.Oligo.motifs">motifs</span></code></td>
            <td>
                  <code><span title="tuple">tuple</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A pair of motifs where the first motif is lowercase (representing the 5' end) and the </p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="dnabyte.oligo.Oligo.type">type</span></code></td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Indicates the type of the oligonucleotide. Defaults to 'single_stranded'.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="ValueError">ValueError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If <code>motifs</code> is provided but is not a valid motif pair (i.e., a tuple of two elements </p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>motifs</code>
            </td>
            <td>
                  <code><span title="tuple">tuple</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A tuple containing a pair of motifs to initialize the oligonucleotide. </p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>








              <details class="mkdocstrings-source">
                <summary>Source code in <code>dnabyte/oligo.py</code></summary>
                <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">  3</span>
<span class="normal">  4</span>
<span class="normal">  5</span>
<span class="normal">  6</span>
<span class="normal">  7</span>
<span class="normal">  8</span>
<span class="normal">  9</span>
<span class="normal"> 10</span>
<span class="normal"> 11</span>
<span class="normal"> 12</span>
<span class="normal"> 13</span>
<span class="normal"> 14</span>
<span class="normal"> 15</span>
<span class="normal"> 16</span>
<span class="normal"> 17</span>
<span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">Oligo</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This class models a single or double stranded oligonucleotide. It is always instanciated as single stranded,</span>
<span class="sd">        from which double stranded oligonucleotides can be created by hybridising it with another oligo.</span>

<span class="sd">        Attributes:</span>
<span class="sd">            motifs (tuple): A pair of motifs where the first motif is lowercase (representing the 5&#39; end) and the </span>
<span class="sd">            second motif is uppercase (representing the 3&#39; end). This attribute is None if no motifs are provided.</span>
<span class="sd">            type (str): Indicates the type of the oligonucleotide. Defaults to &#39;single_stranded&#39;.</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: If `motifs` is provided but is not a valid motif pair (i.e., a tuple of two elements </span>
<span class="sd">            where the first element is lowercase and the second is uppercase).</span>

<span class="sd">        Parameters:</span>
<span class="sd">            motifs (tuple, optional): A tuple containing a pair of motifs to initialize the oligonucleotide. </span>
<span class="sd">            Defaults to None.</span>
<span class="sd">        &quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">motifs</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sequence</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">motifs</span> <span class="o">=</span> <span class="n">motifs</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="nb">type</span>

        <span class="k">if</span> <span class="n">motifs</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">tuple</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s1">&#39;double_stranded&#39;</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">str</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s1">&#39;single_stranded&#39;</span>
        <span class="c1"># TODO: This causes the oligo assignment not to work</span>
        <span class="c1"># Determine the type based on the structure of motifs or sequence</span>
        <span class="c1"># if motifs is not None:</span>
        <span class="c1">#     if isinstance(motifs, list):</span>
        <span class="c1">#         self.type = &#39;double_stranded&#39;</span>
        <span class="c1">#     elif isinstance(motifs, tuple):</span>
        <span class="c1">#         self.type = &#39;single_stranded&#39;</span>
        <span class="c1">#     else:</span>
        <span class="c1">#         raise ValueError(&quot;Invalid motif structure.&quot;)</span>

        <span class="c1"># if sequence is not None:</span>
        <span class="c1">#     if isinstance(sequence, list) and all(isinstance(s, str) for s in sequence):</span>
        <span class="c1">#         self.type = &#39;double_stranded&#39;</span>
        <span class="c1">#     elif isinstance(sequence, str):</span>
        <span class="c1">#         self.type = &#39;single_stranded&#39;</span>
        <span class="c1">#     else:</span>
        <span class="c1">#         raise ValueError(&quot;Invalid sequence structure.&quot;)</span>
        <span class="c1"># else:</span>
        <span class="c1">#     #raise ValueError(&quot;Oligo must be initialized with motifs or sequence.&quot;)</span>
        <span class="c1">#     return None</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">end</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">strand</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">side</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="c1"># If only one argument is provided, it is considered as &#39;side&#39;</span>
        <span class="k">if</span> <span class="n">strand</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">side</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">side</span><span class="p">,</span> <span class="n">strand</span> <span class="o">=</span> <span class="n">strand</span><span class="p">,</span> <span class="kc">None</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s1">&#39;single_stranded&#39;</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">side</span> <span class="o">==</span> <span class="s1">&#39;5&#39;</span> <span class="ow">or</span> <span class="n">side</span> <span class="o">==</span> <span class="s1">&#39;r&#39;</span><span class="p">:</span>
                <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
            <span class="k">elif</span> <span class="n">side</span> <span class="o">==</span> <span class="s1">&#39;3&#39;</span> <span class="ow">or</span> <span class="n">side</span> <span class="o">==</span> <span class="s1">&#39;l&#39;</span><span class="p">:</span>
                <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;f&#39;</span> <span class="ow">and</span> <span class="p">(</span><span class="n">side</span> <span class="o">==</span> <span class="s1">&#39;5&#39;</span> <span class="ow">or</span> <span class="n">side</span> <span class="o">==</span> <span class="s1">&#39;r&#39;</span><span class="p">):</span>
                <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="k">elif</span> <span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;f&#39;</span> <span class="ow">and</span> <span class="p">(</span><span class="n">side</span> <span class="o">==</span> <span class="s1">&#39;3&#39;</span> <span class="ow">or</span> <span class="n">side</span> <span class="o">==</span> <span class="s1">&#39;l&#39;</span><span class="p">):</span>
                <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">elif</span> <span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;r&#39;</span> <span class="ow">and</span> <span class="p">(</span><span class="n">side</span> <span class="o">==</span> <span class="s1">&#39;5&#39;</span> <span class="ow">or</span> <span class="n">side</span> <span class="o">==</span> <span class="s1">&#39;l&#39;</span><span class="p">):</span>
                <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">elif</span> <span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;r&#39;</span> <span class="ow">and</span> <span class="p">(</span><span class="n">side</span> <span class="o">==</span> <span class="s1">&#39;3&#39;</span> <span class="ow">or</span> <span class="n">side</span> <span class="o">==</span> <span class="s1">&#39;r&#39;</span><span class="p">):</span>
                <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>


    <span class="k">def</span><span class="w"> </span><span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">motifs</span> <span class="ow">and</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;motifs&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s1">&#39;single_stranded&#39;</span><span class="p">:</span>
            <span class="n">formatted_motif_1</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot; </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">  &quot;</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">==</span> <span class="mi">1</span> <span class="k">else</span> <span class="sa">f</span><span class="s2">&quot; </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2"> &quot;</span>
            <span class="n">formatted_motif_2</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot; </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2">  &quot;</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">==</span> <span class="mi">1</span> <span class="k">else</span> <span class="sa">f</span><span class="s2">&quot; </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2"> &quot;</span>
            <span class="k">return</span> <span class="s2">&quot;3&#39;:  ( &quot;</span> <span class="o">+</span> <span class="n">formatted_motif_1</span> <span class="o">+</span> <span class="s2">&quot; - &quot;</span> <span class="o">+</span> <span class="n">formatted_motif_2</span> <span class="o">+</span> <span class="s2">&quot; )  :5&#39; </span><span class="se">\n</span><span class="s2">&quot;</span>

        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">motifs</span> <span class="ow">and</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;motifs&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s1">&#39;double_stranded&#39;</span><span class="p">:</span>
            <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
            <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">strand</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">motifs</span><span class="p">):</span>
                <span class="n">line</span> <span class="o">=</span> <span class="s2">&quot;3&#39;:  (&quot;</span> <span class="k">if</span> <span class="n">index</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">else</span> <span class="s2">&quot;5&#39;:  (&quot;</span>

                <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">motif</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">strand</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">motif</span><span class="p">:</span>
                        <span class="n">line</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot; </span><span class="si">{</span><span class="n">motif</span><span class="si">}</span><span class="s2">  &quot;</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">motif</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span> <span class="k">else</span> <span class="sa">f</span><span class="s2">&quot; </span><span class="si">{</span><span class="n">motif</span><span class="si">}</span><span class="s2"> &quot;</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">line</span> <span class="o">+=</span> <span class="s2">&quot;    &quot;</span>
                    <span class="k">if</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">strand</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">:</span>  <span class="c1"># Check if it&#39;s not the last motif</span>
                        <span class="n">line</span> <span class="o">+=</span> <span class="s2">&quot;-&quot;</span>
                <span class="n">line</span> <span class="o">+=</span> <span class="s2">&quot;)  :5&#39;  </span><span class="se">\n</span><span class="s2">&quot;</span> <span class="k">if</span> <span class="n">index</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">else</span> <span class="s2">&quot;)  :3&#39; </span><span class="se">\n</span><span class="s2">&quot;</span>
                <span class="n">output</span> <span class="o">+=</span> <span class="n">line</span>
            <span class="n">output</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="k">return</span> <span class="n">output</span>

        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span> <span class="ow">and</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;sequence&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s1">&#39;single_stranded&#39;</span><span class="p">:</span>
            <span class="k">return</span> <span class="s2">&quot;3&#39;: &quot;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span> <span class="o">+</span> <span class="s2">&quot; :5&#39;&quot;</span>

        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span> <span class="ow">and</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;sequence&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s1">&#39;double_stranded&#39;</span><span class="p">:</span>
            <span class="k">return</span> <span class="s2">&quot;3&#39;: &quot;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot; :5&#39; </span><span class="se">\n</span><span class="s2">5&#39;: &quot;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot; :3&#39;&quot;</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="s2">&quot;Oligo does not have properly defined motifs.&quot;</span>

    <span class="n">wc_pairs</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;A&#39;</span><span class="p">:</span> <span class="s1">&#39;T&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">:</span> <span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">:</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">:</span> <span class="s1">&#39;C&#39;</span><span class="p">}</span>


    <span class="k">def</span><span class="w"> </span><span class="nf">mirror</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns the identical oligonucleotide, where forward and reverse strand are exchanged.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s1">&#39;single_stranded&#39;</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span>
        <span class="k">else</span><span class="p">:</span> 
            <span class="k">return</span> <span class="n">Oligo</span><span class="p">(</span><span class="n">motifs</span><span class="o">=</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">][::</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">][::</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span>
</code></pre></div></td></tr></table></div>
              </details>



<div class="doc doc-children">










<div class="doc doc-object doc-function">


<h2 id="dnabyte.oligo.Oligo.mirror" class="doc doc-heading">
            <code class=" language-python"><span class="n">mirror</span><span class="p">()</span></code>

<a href="#dnabyte.oligo.Oligo.mirror" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents ">

        <p>Returns the identical oligonucleotide, where forward and reverse strand are exchanged.</p>


            <details class="mkdocstrings-source">
              <summary>Source code in <code>dnabyte/oligo.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">mirror</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns the identical oligonucleotide, where forward and reverse strand are exchanged.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s1">&#39;single_stranded&#39;</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span>
    <span class="k">else</span><span class="p">:</span> 
        <span class="k">return</span> <span class="n">Oligo</span><span class="p">(</span><span class="n">motifs</span><span class="o">=</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">][::</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">][::</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div><h3 id="oligopool">OligoPool<a class="headerlink" href="#oligopool" title="Permanent link">&para;</a></h3>


<div class="doc doc-object doc-class">



<a id="dnabyte.oligopool.OligoPool"></a>
    <div class="doc doc-contents first">



        <p>A class to represent a pool of oligonucleotides.</p>
<p>There are two constructors for the OligoPool class:
1. The first constructor takes a list of motifs and generates a pool of oligos with a random copy number for each motif.
The distribution of the copy number is set to be normal, rounded to the next integer. The user can set the mean and
standard deviation of the distribution.</p>
<ol>
<li>The second constructor takes a list of OligoPool objects and generates a new pool by combining the oligos in the
pools of the input OligoPool objects. The user can also set the number of hybridisation events to be performed on the
new pool.</li>
</ol>








              <details class="mkdocstrings-source">
                <summary>Source code in <code>dnabyte/oligopool.py</code></summary>
                <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">  5</span>
<span class="normal">  6</span>
<span class="normal">  7</span>
<span class="normal">  8</span>
<span class="normal">  9</span>
<span class="normal"> 10</span>
<span class="normal"> 11</span>
<span class="normal"> 12</span>
<span class="normal"> 13</span>
<span class="normal"> 14</span>
<span class="normal"> 15</span>
<span class="normal"> 16</span>
<span class="normal"> 17</span>
<span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">OligoPool</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        A class to represent a pool of oligonucleotides.</span>

<span class="sd">        There are two constructors for the OligoPool class:</span>
<span class="sd">        1. The first constructor takes a list of motifs and generates a pool of oligos with a random copy number for each motif.</span>
<span class="sd">        The distribution of the copy number is set to be normal, rounded to the next integer. The user can set the mean and</span>
<span class="sd">        standard deviation of the distribution.</span>

<span class="sd">        2. The second constructor takes a list of OligoPool objects and generates a new pool by combining the oligos in the</span>
<span class="sd">        pools of the input OligoPool objects. The user can also set the number of hybridisation events to be performed on the</span>
<span class="sd">        new pool.</span>
<span class="sd">        &quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">oligo_list</span><span class="p">,</span> <span class="n">mean</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">std_dev</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pool</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">oligo</span> <span class="ow">in</span> <span class="n">oligo_list</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">pool</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pipette</span><span class="p">(</span><span class="n">oligo</span><span class="p">,</span> <span class="n">mean</span><span class="p">,</span> <span class="n">std_dev</span><span class="p">))</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">pipette</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">oligo</span><span class="p">,</span> <span class="n">mean</span><span class="p">,</span> <span class="n">std_dev</span><span class="p">):</span>

        <span class="n">step</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">num_oligos</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">mean</span><span class="p">,</span> <span class="n">std_dev</span><span class="p">)))</span>
        <span class="k">if</span> <span class="n">num_oligos</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">num_oligos</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_oligos</span><span class="p">):</span>
            <span class="n">step</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">oligo</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">step</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">join</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pools</span><span class="p">,</span> <span class="n">mean</span><span class="p">,</span> <span class="n">std_dev</span><span class="p">):</span>

        <span class="c1"># Create a new OligoPool instance</span>
        <span class="n">combined_pool</span> <span class="o">=</span> <span class="n">OligoPool</span><span class="p">([])</span>

        <span class="n">oligo_pools</span> <span class="o">=</span> <span class="p">[</span><span class="n">arg</span> <span class="k">for</span> <span class="n">arg</span> <span class="ow">in</span> <span class="n">pools</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">arg</span><span class="p">,</span> <span class="n">OligoPool</span><span class="p">)]</span>
        <span class="n">oligos</span> <span class="o">=</span> <span class="p">[</span><span class="n">arg</span> <span class="k">for</span> <span class="n">arg</span> <span class="ow">in</span> <span class="n">pools</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">arg</span><span class="p">,</span> <span class="n">Oligo</span><span class="p">)]</span>

        <span class="c1"># Extend the pool with elements from other OligoPool instances</span>
        <span class="k">for</span> <span class="n">pool</span> <span class="ow">in</span> <span class="n">oligo_pools</span><span class="p">:</span>
            <span class="n">combined_pool</span><span class="o">.</span><span class="n">pool</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">pool</span><span class="o">.</span><span class="n">pool</span><span class="p">)</span>

        <span class="c1"># Use the pipette function to append a random copy of the oligos to the pool</span>
        <span class="k">for</span> <span class="n">oligo</span> <span class="ow">in</span> <span class="n">oligos</span><span class="p">:</span>
            <span class="n">combined_pool</span><span class="o">.</span><span class="n">pool</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pipette</span><span class="p">(</span><span class="n">oligo</span><span class="p">,</span> <span class="n">mean</span><span class="p">,</span> <span class="n">std_dev</span><span class="p">))</span>

        <span class="k">return</span> <span class="n">combined_pool</span>


    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">from_oligo_pools</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">oligo_pools</span><span class="p">):</span>
        <span class="n">new_pool_motifs</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">pool</span> <span class="ow">in</span> <span class="n">oligo_pools</span><span class="p">:</span>
            <span class="c1">#new_pool_motifs.extend(pool.oligo_list)</span>
            <span class="n">new_pool_motifs</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">pool</span><span class="o">.</span><span class="n">pool</span><span class="p">)</span>

        <span class="n">new_pool</span> <span class="o">=</span> <span class="bp">cls</span><span class="p">(</span><span class="n">new_pool_motifs</span><span class="p">)</span>
        <span class="n">new_pool</span><span class="o">.</span><span class="n">pool</span> <span class="o">=</span> <span class="p">[</span><span class="n">oligo</span> <span class="k">for</span> <span class="n">pool</span> <span class="ow">in</span> <span class="n">oligo_pools</span> <span class="k">for</span> <span class="n">oligo</span> <span class="ow">in</span> <span class="n">pool</span><span class="o">.</span><span class="n">pool</span><span class="p">]</span>

        <span class="k">return</span> <span class="n">new_pool</span>


    <span class="k">def</span><span class="w"> </span><span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">pool</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">oligo_counts</span> <span class="o">=</span> <span class="p">{}</span>

            <span class="c1"># count occurences of all oligos</span>

            <span class="k">for</span> <span class="n">oligo</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pool</span><span class="p">:</span>
                <span class="n">oligo_key</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">oligo</span><span class="p">)</span>
                <span class="n">oligo_key_mirrored</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">oligo</span><span class="o">.</span><span class="n">mirror</span><span class="p">())</span>

                <span class="k">if</span> <span class="p">(</span><span class="n">oligo_key</span> <span class="ow">in</span> <span class="n">oligo_counts</span><span class="p">):</span>
                    <span class="n">oligo_counts</span><span class="p">[</span><span class="n">oligo_key</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">elif</span> <span class="p">(</span><span class="n">oligo_key_mirrored</span> <span class="ow">in</span> <span class="n">oligo_counts</span><span class="p">):</span>
                    <span class="n">oligo_counts</span><span class="p">[</span><span class="n">oligo_key_mirrored</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">oligo_counts</span><span class="p">[</span><span class="n">oligo_key</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>

            <span class="c1"># sort oligos by count, descending</span>
            <span class="n">sorted_oligos</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">oligo_counts</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">item</span><span class="p">:</span> <span class="n">item</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>

            <span class="c1"># generate output string</span>
            <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
            <span class="k">for</span> <span class="n">oligo_key</span><span class="p">,</span> <span class="n">count</span> <span class="ow">in</span> <span class="n">sorted_oligos</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="n">output</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">oligo_key</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">count</span><span class="si">}</span><span class="s2"> </span><span class="se">\n</span><span class="s2"> </span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="k">return</span> <span class="n">output</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="s2">&quot;Oligo pool is empty.&quot;</span>


    <span class="k">def</span><span class="w"> </span><span class="nf">hybridise</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n</span><span class="p">,</span> <span class="n">library</span><span class="p">,</span> <span class="n">info</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>

        <span class="c1"># set default value for n</span>
        <span class="k">if</span> <span class="n">n</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">n</span> <span class="o">=</span> <span class="mi">10</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pool</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pool</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>  <span class="c1"># Ensure there are at least two oligos to hybridize</span>
                <span class="c1"># Randomly pick two distinct oligos</span>
                <span class="n">oligo_A</span><span class="p">,</span> <span class="n">oligo_B</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pool</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>

                <span class="c1"># Pair the selected oligos</span>
                <span class="n">hybridised_oligo</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pair</span><span class="p">(</span><span class="n">oligo_A</span><span class="p">,</span> <span class="n">oligo_B</span><span class="p">,</span><span class="n">library</span><span class="p">)</span>

                <span class="c1"># Check if the hybridisation was successful</span>
                <span class="k">if</span> <span class="n">hybridised_oligo</span><span class="o">.</span><span class="n">motifs</span><span class="p">:</span>
                    <span class="c1"># Remove the original oligos from the pool</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">pool</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">oligo_A</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">pool</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">oligo_B</span><span class="p">)</span>

                    <span class="c1"># Append the new hybridised oligo to the pool</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">pool</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">hybridised_oligo</span><span class="p">)</span>

                    <span class="c1">#if info:  # Optionally print information about the hybridisation event</span>


        <span class="k">return</span> <span class="bp">self</span>


    <span class="k">def</span><span class="w"> </span><span class="nf">pair</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">oligo_A</span><span class="p">,</span> <span class="n">oligo_B</span><span class="p">,</span><span class="n">library</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">motif_dict</span><span class="o">=</span><span class="n">library</span><span class="o">.</span><span class="n">dictmotives</span>

        <span class="k">if</span> <span class="n">oligo_A</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s1">&#39;single_stranded&#39;</span> <span class="ow">and</span> <span class="n">oligo_B</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s1">&#39;single_stranded&#39;</span><span class="p">:</span>  

            <span class="k">if</span> <span class="n">complement</span><span class="p">(</span><span class="n">oligo_A</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;3&#39;</span><span class="p">),</span><span class="bp">self</span><span class="o">.</span><span class="n">motif_dict</span><span class="p">)</span> <span class="o">==</span> <span class="n">oligo_B</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;3&#39;</span><span class="p">):</span>
                <span class="n">new_motifs</span> <span class="o">=</span> <span class="p">((</span><span class="kc">None</span><span class="p">,</span> <span class="n">oligo_A</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">oligo_A</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span> <span class="p">(</span><span class="n">oligo_B</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">oligo_B</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="kc">None</span><span class="p">))</span>

            <span class="k">elif</span> <span class="n">complement</span><span class="p">(</span><span class="n">oligo_A</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;5&#39;</span><span class="p">),</span><span class="bp">self</span><span class="o">.</span><span class="n">motif_dict</span><span class="p">)</span> <span class="o">==</span> <span class="n">oligo_B</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;5&#39;</span><span class="p">):</span>
                <span class="n">new_motifs</span> <span class="o">=</span> <span class="p">((</span><span class="n">oligo_A</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">oligo_A</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="kc">None</span><span class="p">),</span> <span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="n">oligo_B</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">oligo_B</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>                

            <span class="k">else</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">Oligo</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>

            <span class="n">new_oligo</span> <span class="o">=</span> <span class="n">Oligo</span><span class="p">(</span><span class="n">new_motifs</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">new_oligo</span>


        <span class="k">if</span> <span class="n">oligo_A</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s1">&#39;double_stranded&#39;</span> <span class="ow">and</span> <span class="n">oligo_B</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s1">&#39;double_stranded&#39;</span><span class="p">:</span>  
            <span class="k">if</span> <span class="n">oligo_A</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;f&#39;</span><span class="p">,</span> <span class="s1">&#39;5&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">oligo_B</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="s1">&#39;5&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">oligo_A</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;f&#39;</span><span class="p">,</span> <span class="s1">&#39;5&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="n">complement</span><span class="p">(</span><span class="n">oligo_B</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="s1">&#39;5&#39;</span><span class="p">),</span><span class="bp">self</span><span class="o">.</span><span class="n">motif_dict</span><span class="p">):</span>

                <span class="n">new_motifs</span> <span class="o">=</span> <span class="p">(</span><span class="n">oligo_A</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">oligo_B</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">:],</span>    <span class="c1"># forward strand</span>
                                <span class="n">oligo_A</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">oligo_B</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>  <span class="c1"># reverse strand</span>

            <span class="c1"># case: dsA forward 5&#39; sticky end binds to dsB forward 5&#39; sticky end</span>
            <span class="k">elif</span> <span class="n">oligo_A</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;f&#39;</span><span class="p">,</span> <span class="s1">&#39;5&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">oligo_B</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;f&#39;</span><span class="p">,</span> <span class="s1">&#39;5&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">oligo_A</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;f&#39;</span><span class="p">,</span> <span class="s1">&#39;5&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="n">complement</span><span class="p">(</span><span class="n">oligo_B</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;f&#39;</span><span class="p">,</span> <span class="s1">&#39;5&#39;</span><span class="p">),</span><span class="bp">self</span><span class="o">.</span><span class="n">motif_dict</span><span class="p">):</span>

                <span class="n">new_motifs</span> <span class="o">=</span> <span class="p">(</span><span class="n">oligo_A</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">oligo_B</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="o">-</span><span class="mi">2</span><span class="p">::</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span>        <span class="c1"># forward strand</span>
                                <span class="n">oligo_A</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">oligo_B</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">][::</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>    <span class="c1"># reverse strand</span>

            <span class="c1"># case: the sticky end is on the reverse strand on the right and on the forward strand on the left</span>
            <span class="k">elif</span> <span class="n">oligo_A</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="s1">&#39;3&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">oligo_B</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;f&#39;</span><span class="p">,</span> <span class="s1">&#39;3&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">oligo_A</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="s1">&#39;3&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="n">complement</span><span class="p">(</span><span class="n">oligo_B</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;f&#39;</span><span class="p">,</span> <span class="s1">&#39;3&#39;</span><span class="p">),</span><span class="bp">self</span><span class="o">.</span><span class="n">motif_dict</span><span class="p">):</span>

                <span class="n">new_motifs</span> <span class="o">=</span> <span class="p">(</span><span class="n">oligo_A</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">oligo_B</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>    <span class="c1"># forward strand</span>
                                <span class="n">oligo_A</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">oligo_B</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">1</span><span class="p">:])</span>        <span class="c1"># reverse strand</span>

            <span class="c1"># case: the sticky ends are on the reverse strands on the right</span>

            <span class="k">elif</span> <span class="n">oligo_A</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="s1">&#39;3&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">oligo_B</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="s1">&#39;3&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">oligo_A</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="s1">&#39;3&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="n">complement</span><span class="p">(</span><span class="n">oligo_B</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="s1">&#39;3&#39;</span><span class="p">),</span><span class="bp">self</span><span class="o">.</span><span class="n">motif_dict</span><span class="p">):</span>

                <span class="n">new_motifs</span> <span class="o">=</span> <span class="p">(</span><span class="n">oligo_A</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">][:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">oligo_B</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">][::</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span>    <span class="c1"># forward strand</span>
                                <span class="n">oligo_A</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">oligo_B</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">][:</span><span class="o">-</span><span class="mi">1</span><span class="p">][::</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>        <span class="c1"># reverse strand</span>

            <span class="c1"># case: the sticky end is on the forward strand on the left and on the reverse strand on the right</span>
            <span class="k">elif</span> <span class="n">oligo_A</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;f&#39;</span><span class="p">,</span> <span class="s1">&#39;3&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">oligo_B</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="s1">&#39;3&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">oligo_A</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;f&#39;</span><span class="p">,</span> <span class="s1">&#39;3&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="n">complement</span><span class="p">(</span><span class="n">oligo_B</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="s1">&#39;3&#39;</span><span class="p">),</span><span class="bp">self</span><span class="o">.</span><span class="n">motif_dict</span><span class="p">):</span>

                <span class="n">new_motifs</span> <span class="o">=</span> <span class="p">(</span> <span class="n">oligo_B</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">oligo_A</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>    <span class="c1"># forward strand</span>
                                <span class="n">oligo_B</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">oligo_A</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">1</span><span class="p">:])</span>  <span class="c1"># reverse strand</span>

            <span class="c1"># case: the sticky ends are on the forward strand on the left</span>
            <span class="k">elif</span> <span class="n">oligo_A</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;f&#39;</span><span class="p">,</span> <span class="s1">&#39;3&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">oligo_B</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;f&#39;</span><span class="p">,</span> <span class="s1">&#39;3&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">oligo_A</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;f&#39;</span><span class="p">,</span> <span class="s1">&#39;3&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="n">complement</span><span class="p">(</span><span class="n">oligo_B</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;f&#39;</span><span class="p">,</span> <span class="s1">&#39;3&#39;</span><span class="p">),</span><span class="bp">self</span><span class="o">.</span><span class="n">motif_dict</span><span class="p">):</span>

                <span class="n">new_motifs</span> <span class="o">=</span> <span class="p">(</span> <span class="n">oligo_B</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">1</span><span class="p">:][::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">oligo_A</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>    <span class="c1"># forward strand</span>
                                <span class="n">oligo_B</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">][::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">oligo_A</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">1</span><span class="p">:])</span>  <span class="c1"># reverse strand</span>

            <span class="c1"># case: the sticky end is on the reverse strand on the left and the forward strand on the right</span>
            <span class="k">elif</span> <span class="n">oligo_A</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="s1">&#39;5&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">oligo_B</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;f&#39;</span><span class="p">,</span> <span class="s1">&#39;5&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">oligo_A</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="s1">&#39;5&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="n">complement</span><span class="p">(</span><span class="n">oligo_B</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;f&#39;</span><span class="p">,</span> <span class="s1">&#39;5&#39;</span><span class="p">),</span><span class="bp">self</span><span class="o">.</span><span class="n">motif_dict</span><span class="p">):</span>

                <span class="n">new_motifs</span> <span class="o">=</span> <span class="p">(</span><span class="n">oligo_B</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">oligo_A</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">:],</span>    <span class="c1"># forward strand</span>
                                <span class="n">oligo_B</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">][:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">oligo_A</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>  <span class="c1"># reverse strand</span>

            <span class="c1"># case: the sticky ends are on the reverse strand on the left</span>
            <span class="k">elif</span> <span class="n">oligo_A</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="s1">&#39;5&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">oligo_B</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="s1">&#39;5&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">oligo_A</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="s1">&#39;5&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="n">complement</span><span class="p">(</span><span class="n">oligo_B</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="s1">&#39;5&#39;</span><span class="p">),</span><span class="bp">self</span><span class="o">.</span><span class="n">motif_dict</span><span class="p">):</span>

                <span class="n">new_motifs</span> <span class="o">=</span> <span class="p">(</span><span class="n">oligo_B</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">][::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">oligo_A</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">:],</span>    <span class="c1"># forward strand</span>
                                <span class="n">oligo_B</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">][:</span><span class="mi">0</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">oligo_A</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>  <span class="c1"># reverse strand</span>

            <span class="k">else</span><span class="p">:</span> 
                <span class="k">return</span> <span class="n">Oligo</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span> 

            <span class="k">return</span> <span class="n">Oligo</span><span class="p">(</span><span class="n">new_motifs</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span> <span class="c1"># case where one oligo is single stranded and the other is double stranded</span>

            <span class="k">if</span> <span class="n">oligo_A</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s1">&#39;double_stranded&#39;</span> <span class="ow">and</span> <span class="n">oligo_B</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s1">&#39;single_stranded&#39;</span><span class="p">:</span> 
                <span class="n">ds</span> <span class="o">=</span> <span class="n">oligo_A</span>
                <span class="n">ss</span> <span class="o">=</span> <span class="n">oligo_B</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">ds</span> <span class="o">=</span> <span class="n">oligo_B</span>
                <span class="n">ss</span> <span class="o">=</span> <span class="n">oligo_A</span>

            <span class="c1"># case: double strand forward 5&#39; sticky end binds to ss 5&#39; sticky end</span>
            <span class="k">if</span> <span class="n">ds</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;f&#39;</span><span class="p">,</span> <span class="s1">&#39;5&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">ds</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;f&#39;</span><span class="p">,</span> <span class="s1">&#39;5&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="n">complement</span><span class="p">(</span><span class="n">ss</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;5&#39;</span><span class="p">),</span><span class="bp">self</span><span class="o">.</span><span class="n">motif_dict</span><span class="p">):</span>

                <span class="n">new_motifs</span><span class="o">=</span><span class="p">(</span><span class="n">ds</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="p">(</span><span class="kc">None</span><span class="p">,),</span>                                 <span class="c1"># forward strand</span>
                                <span class="n">ds</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">][:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="p">(</span><span class="n">ss</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">ss</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>   <span class="c1"># reverse strand</span>

            <span class="c1"># case: ds reverse 3&#39; sticky end binds to ss 3&#39; sticky end</span>
            <span class="k">elif</span> <span class="n">ds</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="s1">&#39;3&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">ds</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="s1">&#39;3&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="n">complement</span><span class="p">(</span><span class="n">ss</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;3&#39;</span><span class="p">),</span><span class="bp">self</span><span class="o">.</span><span class="n">motif_dict</span><span class="p">):</span>

                <span class="n">new_motifs</span> <span class="o">=</span> <span class="p">(</span><span class="n">ds</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="p">(</span><span class="n">ss</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">ss</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span>    <span class="c1"># forward strand</span>
                                <span class="n">ds</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="p">(</span><span class="kc">None</span><span class="p">,))</span>                             <span class="c1"># reverse strand</span>

            <span class="c1"># case: ds forward 3&#39; sticky end binds to ss 5&#39; sticky end</span>
            <span class="k">elif</span> <span class="n">ds</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;f&#39;</span><span class="p">,</span> <span class="s1">&#39;3&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">ds</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;f&#39;</span><span class="p">,</span> <span class="s1">&#39;3&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="n">complement</span><span class="p">(</span><span class="n">ss</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;3&#39;</span><span class="p">),</span><span class="bp">self</span><span class="o">.</span><span class="n">motif_dict</span><span class="p">):</span>  

                <span class="n">new_motifs</span> <span class="o">=</span> <span class="p">((</span><span class="kc">None</span><span class="p">,)</span> <span class="o">+</span> <span class="n">ds</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>                               <span class="c1"># forward strand</span>
                                <span class="p">(</span><span class="n">ss</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">ss</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">+</span> <span class="n">ds</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">1</span><span class="p">:])</span>    <span class="c1"># reverse strand</span>

            <span class="c1"># case: ds reverse 5&#39; sticky end binds to ss 5&#39; sticky end</span>
            <span class="k">elif</span> <span class="n">ds</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="s1">&#39;5&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">ds</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="s1">&#39;5&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="n">complement</span><span class="p">(</span><span class="n">ss</span><span class="o">.</span><span class="n">end</span><span class="p">(</span><span class="s1">&#39;5&#39;</span><span class="p">),</span><span class="bp">self</span><span class="o">.</span><span class="n">motif_dict</span><span class="p">):</span>

                <span class="n">new_motifs</span> <span class="o">=</span> <span class="p">((</span><span class="n">ss</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">ss</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">+</span> <span class="n">ds</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">:],</span>      <span class="c1"># forward strand</span>
                                <span class="p">(</span><span class="kc">None</span><span class="p">,)</span> <span class="o">+</span> <span class="n">ds</span><span class="o">.</span><span class="n">motifs</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>                             <span class="c1"># reverse strand</span>

            <span class="k">else</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">Oligo</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>

            <span class="k">return</span> <span class="n">Oligo</span><span class="p">(</span><span class="n">new_motifs</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



<div class="doc doc-children">












  </div>

    </div>

</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        
<div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://github.com/Eko-Refugium" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M202.1 328.7c0 20.9-10.9 55.1-36.7 55.1s-36.7-34.2-36.7-55.1 10.9-55.1 36.7-55.1 36.7 34.2 36.7 55.1M496 278.2c0 31.9-3.2 65.7-17.5 95-37.9 76.6-142.1 74.8-216.7 74.8-75.8 0-186.2 2.7-225.6-74.8-14.6-29-20.2-63.1-20.2-95 0-41.9 13.9-81.5 41.5-113.6-5.2-15.8-7.7-32.4-7.7-48.8 0-21.5 4.9-32.3 14.6-51.8 45.3 0 74.3 9 108.8 36 29-6.9 58.8-10 88.7-10 27 0 54.2 2.9 80.4 9.2 34-26.7 63-35.2 107.8-35.2 9.8 19.5 14.6 30.3 14.6 51.8 0 16.4-2.6 32.7-7.7 48.2 27.5 32.4 39 72.3 39 114.2m-64.3 50.5c0-43.9-26.7-82.6-73.5-82.6-18.9 0-37 3.4-56 6-14.9 2.3-29.8 3.2-45.1 3.2-15.2 0-30.1-.9-45.1-3.2-18.7-2.6-37-6-56-6-46.8 0-73.5 38.7-73.5 82.6 0 87.8 80.4 101.3 150.4 101.3h48.2c70.3 0 150.6-13.4 150.6-101.3m-82.6-55.1c-25.8 0-36.7 34.2-36.7 55.1s10.9 55.1 36.7 55.1 36.7-34.2 36.7-55.1-10.9-55.1-36.7-55.1"/></svg>
    </a>
  
    
    
    
    
      
      
    
    <a href="https://www.linkedin.com/company/96306392/" target="_blank" rel="noopener" title="www.linkedin.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3M135.4 416H69V202.2h66.5V416zM102.2 96a38.5 38.5 0 1 1 0 77 38.5 38.5 0 1 1 0-77m282.1 320h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9z"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "..", "features": ["icons"], "search": "../assets/javascripts/workers/search.7a47a382.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.e71a0d61.min.js"></script>
      
    
  </body>
</html>